---
title: "Investigation of the role of PHD2 atherogenesis derived from myeloid lineage in-vivo"
author: "Javier Perales-PatÃ³n - javier.perales@bioquant.uni-heidelberg.de - ORCID: 0000-0003-0780-6683" 
output: github_document
---

> Herein we focus on myeloid cells to study their response to PHD2cKO and its relation with atherogenesis in-vivo. 
From the in-vitro experiments, we showed that PHD2-KO in macrophages leads to Bnip3 apoptosis which aggravates atherogenesis. 
Actually we observed that PHD2-KO induction of Bnip3 apoptosis in Macrophages leads to a cross-talk with fibroblasts, which is 
associated with increased ECM production in atherosclerotic plaques. Moreover, this event could be reduced by depleting
Bnip3 in Macrophages (upstream). To investigate this further, we also have profiled bulk in-vitro data of this cellular 
cross-talk and characterized the mechanism. 
We predicted that Spp1 released by macrophage might be the ligand involved in this interaction with fibroblasts. 
Here, the data is derived from the in-vivo model which carries PHD2-KO in myeloid lineage. Thus we analyse 
the PHD2KO-Bnip3-Spp1 axis, and look into which cells might be the source of Spp1 in-vivo via characterization of myeloid 
leukocytes (incl. macrophages) in this context.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Setup
We define a random seed number for reproducibility, file structure for the output, 
and load essential libraries

### Environment
```{r env}
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./03_myeloidlineage_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png','tiff'))
knitr::opts_chunk$set(dpi=300)
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);
```

### Load libraries
```{r}
suppressPackageStartupMessages(require(Seurat))
suppressPackageStartupMessages(require(cowplot))
suppressPackageStartupMessages(require(ggplot2))
suppressPackageStartupMessages(require(purrr))
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(require(openxlsx))
suppressPackageStartupMessages(require(GSEABase))
suppressPackageStartupMessages(require(ComplexHeatmap))
suppressPackageStartupMessages(require(AUCell))
suppressPackageStartupMessages(require(rcompanion))
source("../src/graphics.R")
source("../src/seurat_fx.R")
source("../src/wilcox_fx.R")
```

## Load data
Read the Seurat Object from second step.
```{r load_seurat}
# Input data
sobj <- "./02_identity_output/data/S.rds"
# Read 
if(file.exists(sobj)) {
	S <- readRDS(sobj)
} else {
	stop("ERROR: Seurat object does not exist. Run 02.rmd to generate it.")
}
```

We also load the ortholog information for functional analysis. We will use it for
the allocation of cell cycle phases in single cells.
```{r cc_mmu}
# Lib of hsa2mmu orthologs
mmu2hsa <- readRDS("../data/Gene_annotation/hgnc2mgi_hsa2mmu.rds")
# Cycling genes from human to mouse
CC.genes <- lapply(cc.genes, function(vect) unlist(sapply(vect, function(z) mmu2hsa[[z]][1])))
# Allocation
S <- CellCycleScoring(S, s.features=CC.genes$s.genes, g2m.features=CC.genes$g2m.genes)
```

## Myeloid lineage
Myieloid lineage is composed by megakaryocyte (trombocytes), erytrocyte, mast cells, 
basophil, neutrophils, eosinophils, monocytes and macrophages. So far we have identified
the following populations of these in our map:

```{r}
myeloid <- c("Megakaryocyte", "Thrombocyte", "Erythrocyte", "Mast cell",
	     "Basophil", "Neutrophil", "Eosinophil", "Monocyte", "Macrophage")
grep(paste(myeloid, collapse="|"), levels(S), value=TRUE)
```

We use the marker "Lyz2" to highlight myeloid cells. Actually, the small cluster No. 8, 
named previously as "n.a", is a cluster of immune cells expressingn Lyz2 pending to be identified.
They also express other markers of leukocytes as shown in previous step. So it must be a myeloid leucocyte as well. 
In the following plot it is shown that Monocytes, Neutrophils, DC, Macrophages and that small cluster 
express the myeloid marker.

The violin plot of Lyz2 expression show that the aforementioned clusters over-express
the myeloid marker.

```{r expression_Lyz2, fig.width=12, fig.height=8, warning=FALSE}
umap_Lyz2 <- FeaturePlot(S, feature="Lyz2", label=TRUE)
vln_Lyz2 <- VlnPlot(S, feature="Lyz2") + NoLegend()
plot_grid(umap_Lyz2, vln_Lyz2)
```

Then we subset the dataset by these selected populations.
```{r}
# We save the initial annotation for the records
S$Annotation.1 <- Idents(S)
# We subset
M <- S[, Idents(S) %in% c("Mixed_Neutrophil-DC-Monocytes", "Macrophage", "n.a.")]
```

Zernecke et al 2020 is a meta-analysis review of Leukocytes in (mouse) atherosclerotic plaques.
We show that macrophage express all markers of macrophage classes described in Zernecke et al 2020.
While the mixed cluster of Neutrophils, DC and monocytes do so with the corresponding markers of these clases.
The third small cluster is not clear which cell type represent. Thus, it remains as n.a. (Not Assigned).

```{r heatmap_Zernecke_initML, fig.width=14, fig.height=10, dpi=300, warning=FALSE}
# Top10 markers for each Myeloid Leukocytes found in Meta-analysis review Zernecke et al. 2020
ML.markers <- getGmt("../data/markers/leukocytes_Zernecke2020.gmt")
ML.markers <- ML.markers[c(grep("_Mac", names(ML.markers)),
			   grep("_Mac", names(ML.markers), invert=TRUE))]

hp <- DoHeatmap3(SeuratObject=M, GSC=ML.markers, assay="RNA", res="Idents", 
	   row_names_size=8, column_title_size=0,
	   show_hr=FALSE, legend_nrow=2, fontfamily=fontTXT) 
draw(hp, heatmap_legend_side="right", annotation_legend_side="bottom")
```

## Full-identification of Myeloid lineage

> For the full identification of myeloid leukocytes, we recluster these cells in a second iteration of
unsupervised clustering with higher resolution (res=0.5). Then we map again myeloid leukocytes markers. 
During the preparation of the data analysis, Zernecke et al 2020 has published a comprehensive catalog 
of markers for myeloid leukocytes in (mouse) atherosclerotic plaques. Thus from Zernecke et al 2020 
to aid with the annotation.

We split the cells by dataset to perform a full workflow of data integration, pca and clustering with
Seurat.

```{r reclustering_MyeloidLeukocytes}
M.list <- SplitObject(M,
		      split.by="orig.ident")

for(i in names(M.list)){
	M.list[[i]] <- NormalizeData(M.list[[i]], verbose=FALSE)
	M.list[[i]] <- ScaleData(M.list[[i]], verbose=FALSE)
	M.list[[i]] <- FindVariableFeatures(M.list[[i]], 
					   selection.method = "vst",
					   nfeatures = 2000, verbose = FALSE)

}
M.anchors <- FindIntegrationAnchors(object.list = M.list, dims = 1:30)
M.i <- IntegrateData(anchorset = M.anchors, dims = 1:30)

# Clean metadata
M.i@meta.data <- M.i@meta.data[, grep("snn_res", colnames(M.i@meta.data), invert=TRUE)]
Project(M.i) <- "Myeloid_Leukocytes"
if(is.character(M.i$stim)) M.i$stim <- factor(M.i$stim, levels=c("WT","PHD2cKO")) ;

if(all(dim(M.i@assays$RNA@scale.data) == c(0,0))) {
M.i@assays$RNA@scale.data <- cbind(M.list[[1]]@assays$RNA@scale.data,
				   M.list[[2]]@assays$RNA@scale.data)[,colnames(M.i)] 
}

DefaultAssay(M.i) <- "integrated"

# Run the standard workflow for visualization and clustering
M.i <- ScaleData(M.i, verbose = FALSE)
M.i <- RunPCA(M.i, npcs = 30, verbose = FALSE)
# ElbowPlot(M.i)
M.i <- RunUMAP(M.i, reduction = "pca", dims = 1:15)
```

It seems that both conditions present all types of myeloid leukocytes.
```{r umap_Myeloid_condition, fig.width=5, fig.height=5, dpi=300}
DimPlot(M.i, reduction="umap", group.by="orig.ident", cols=c("red","grey")) +
	guides(colour = guide_legend(override.aes = list(size=8))) + 
	theme(legend.position="bottom",
	      axis.title = element_text(size=16),
	      axis.text= element_text(size=16))
```

Second iteration of unsupervised clustering to find even major differences in this cell 
population of myeloid leukocytes.
```{r umap_myeloid_clusters, fig.width=5, fig.height=5, dpi=300}
set.seed(1234)
# Find clusters
M.i <- FindNeighbors(M.i, dims = 1:20)
M.i <- FindClusters(M.i, resolution = 0.5)
table(Idents(M.i))

DefaultAssay(M.i) <- "RNA"
```

These are the major distinct cell populations
```{r umap_Myeloid_clusters, fig.width=5, fig.height=5, dpi=300}
DimPlot(M.i, reduction="umap") +
	guides(colour = guide_legend(override.aes = list(size=8), nrow=2)) + 
	theme(legend.position="bottom",
	      axis.title = element_text(size=16),
	      axis.text= element_text(size=16))
```

There is one cluster that are proliferating cells.
```{r vln_myeloid_cellcycle, fig.with=7, fig.height=5, dpi=300}
VlnPlot(M.i, feature=c("S.Score", "G2M.Score"))
```

We confirm that these are Myeloid Leukocytes (Lyz2, Cd68) and discard other cellular 
lineages (epithelial and endothelial, Myh11, Cdh5).

```{r umap_Myeloid_candidates, fig.width=10, fig.height=8, dpi=300, warning=FALSE}
tm <- theme(title = element_text(size=26, family=fontTXT),
        axis.title = element_text(size=20, family=fontTXT),
        axis.text = element_text(size=20, family=fontTXT),
        legend.text = element_text(size=18, family=fontTXT)
        )
p1 <- FeaturePlot(M.i, features=c("Lyz2")) + tm
p2 <- FeaturePlot(M.i, features=c("Cd68")) + tm
p3 <- FeaturePlot(M.i, features=c("Myh11")) + tm
p4 <- FeaturePlot(M.i, features=c("Cdh5")) + tm
## Other candidates, immuno markers
# FeaturePlot(M.i, features=c("Cd3e")) + tm
# FeaturePlot(M.i, features=c("Ptprc")) + tm
CombinePlots(list(p1,p2,p3,p4), ncol = 2)
```



Based on markers from next chunks of code, we assign identities to the 
different cluster of cells. Please note that we have already assigned them (restrospectively)
for corcondance with final plots, but the data shows progresively that these annotations 
are correct.

```{r myeloid_rids}
M.i <- RenameIdents(M.i, c("0"="Resident-Mac_1",
			   "1"="TREM2hi-Mac",
			   "2"="Inflammatory-Mac",
			   "3"="Monocyte",
			   "4"="Cavity-Mac",
			   "5"="Neutrophil",
			   "6"="MoDC",
			   "7"="Proliferating",
			   "8"="Resident-Mac_2",
			   "9"="n.a.",
			   "10"="Mature-DC",
			   "11"="IFNIC-Mac"))
# Sort them by, 1st Macrophages, then the rest, finally the odd ones
last_ids<- c("n.a.","Proliferating")
ids <- c(sort(grep("-Mac", levels(M.i), value=TRUE)),
	 setdiff(sort(grep("-Mac", levels(M.i), value=TRUE, invert=TRUE)),
		 last_ids),
	 last_ids)

Idents(M.i) <- factor(as.character(Idents(M.i)),
		      levels=ids)
# We show the total number of cells
ncol(M.i)
# And the sample size of each group
table(Idents(M.i))
# We save the sample size
write.table(as.data.frame(table(Cluster=Idents(M.i))),
	    file=paste0(DATADIR,"/nCell_per_cluster.csv"),
	    sep=",", row.names=FALSE, col.names=TRUE, quote=FALSE)
```

These cell assignment of identities are supported by Zernecke et al 2020 - a meta-analysis of
leukocytes in (mouse) atherosclerotic plaques.

```{r heatmap_Zernecke_MyeloidLeukocytes, fig.width=14, fig.height=10, dpi=300, warning=FALSE}
# Top10 markers for each Myeloid Leukocytes found in Meta-analysis review Zernecke et al. 2020
ML.markers <- getGmt("../data/markers/leukocytes_Zernecke2020.gmt")
ML.markers <- ML.markers[c(grep("_Mac", names(ML.markers)),
			   grep("_Mac", names(ML.markers), invert=TRUE))]

hp <- DoHeatmap3(SeuratObject=M.i, GSC=ML.markers, assay="RNA", res="Idents", 
	   row_names_size=8, column_title_size=0,
	   show_hr=FALSE, legend_nrow=2, fontfamily=fontTXT) 
draw(hp, heatmap_legend_side="right", annotation_legend_side="bottom")
```

We remake that heatmap but making it more compact (without showing gene names). The genes are shown
above and these are described in [leukocytes_Zernecke2020.gmt](../data/markers/leukocytes_Zernecke2020.gmt).

```{r heatmap_Zernecke_MyeloidLeukocytes_small, fig.width=9, fig.height=8, dpi=300, warning=FALSE}
hp <- DoHeatmap3(SeuratObject=M.i, GSC=ML.markers, assay="RNA", res="Idents", 
	   row_names_size=0, column_title_size=0,
	   show_hr=FALSE, legend_nrow=3, fontfamily=fontTXT) 
draw(hp, heatmap_legend_side="right", annotation_legend_side="bottom")
```

There is one population of macrophages not very clear whether they are inflammatory or not.
So we extend the marker list to all differentially up-regulated genes reported in Zernecke et al 2020.

```{r inflammatory_Zernecke_MyeloidLeukocytes, fig.width=14, fig.height=10, dpi=300, warning=FALSE}
inflam <- read.table("../data/markers/inflammatory-Mac_DEGs_Zernecke2020.tsv", 
		     sep="\t", header=TRUE, stringsAsFactors=FALSE)

inflam <- inflam[inflam$p_val_adj < 0.05, ]

inflam <- GeneSetCollection(GeneSet(inflam$gene, setName="Inflamm.Mac"))
hp <- DoHeatmap3(SeuratObject=M.i, GSC=inflam, assay="RNA", res="Idents", 
	   row_names_size=0, column_title_size=0,
	   show_hr=FALSE, legend_nrow=2, fontfamily=fontTXT) 
draw(hp, heatmap_legend_side="right", annotation_legend_side="bottom")
```

```{r vln_aucell_inflam_macrophages, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
# Build rankings for AUC
cells_rankings <- AUCell_buildRankings(as.matrix(M.i@assays$RNA@data))
# Use top N (e.g. 50) to calculate enrichment at single-cell level
cells_AUC <- AUCell_calcAUC(lapply(geneIds(inflam), function(gs) head(gs,50)), 
			cells_rankings, 
			 aucMaxRank=nrow(cells_rankings)*0.05)
# Explore thresholds
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=FALSE, assign=TRUE) 
# Get AUCell scores  
AUC.sign <- getAUC(cells_AUC)[1,]
stopifnot(colnames(M.i) == names(AUC.sign))
M.i$inflam.Mac<- AUC.sign
rm(AUC.sign, cells_rankings, cells_AUC, cells_assignment)
VlnPlot(M.i, feature="inflam.Mac")
```

> We conclude that the remaining group of macrophages must be "inflammatory macrophages",
since these are the only one class of macrophages that remains missing, and the group is
expressing more consistently and highly differentially expressed genes of this class.


The final UMAP is shown. We still have one very small population of cells that is
not assigned (n.a.) and a cluster of cells that are proliferating.

```{r umap_Myeloid_cluster_assign, fig.width=8, fig.height=8.5, dpi=300}
DimPlot(M.i, reduction="umap", label=FALSE) +
	guides(colour = guide_legend(override.aes = list(size=6), nrow=3)) + 
	coord_cartesian(clip="off") +
	theme(legend.position="bottom",
	      axis.title = element_text(size=16),
	      axis.text= element_text(size=16)) 

DimPlot(M.i, reduction="umap", label=TRUE, label.size=6) +
	guides(colour = guide_legend(override.aes = list(size=6), nrow=3)) + 
	coord_cartesian(clip="off") +
	theme(legend.position="bottom",
	      axis.title = element_text(size=16),
	      axis.text= element_text(size=16)) 
```

We confirm again the proliferating one.

```{r vln_myeloid_cellcycle_assign, fig.with=7, fig.height=5, dpi=300}
VlnPlot(M.i, feature=c("S.Score", "G2M.Score"))
```

Finally, we show their proportion in each condition. This is informative, but we cannot
make statistical contrasts on this direction since these are two sequencing runs (pooled mice), 
so we do not have more than 1 "replicate" in our sample populations.


```{r bp_ML_clusters, fig.width=4.5, fig.height=2, dpi=300, warning=FALSE}
# Cell freqs
cell_pop <- table(M.i$orig.ident,Idents(M.i))
print(cell_pop)

# Cell Proportions
cell_pop_perc <- (sweep(cell_pop,MARGIN = 1, STATS = rowSums(cell_pop), FUN = "/")) * 100
cell_pop_perc2 <- reshape2::melt(cell_pop_perc)
cell_pop_perc2$Var1 <- relevel(cell_pop_perc2$Var1, ref = "WT")


ggplot(cell_pop_perc2, aes(x= Var1, fill=Var2, y=value)) +
    geom_col(position = position_stack(reverse=TRUE)) + 
#          scale_fill_manual(values = cols[levels(cell_pop_perc2$Var2)]) +
  scale_y_continuous(position="right") +
  labs(y="Percentage of cells (%)") + theme_bw() + coord_flip() +
  theme(legend.title = element_blank(),
	legend.text = element_text(size=7, family=fontTXT),
    legend.key.size = unit(0.5, "cm"),
    legend.position = "bottom",
        axis.title.x = element_text(size=18, color = "black", family=fontTXT),
    axis.title.y = element_blank(),
        axis.text.x = element_text(size=14, color = "black", family=fontTXT),
        axis.text.y = element_text(size=12, color = "black", family=fontTXT),
        panel.grid = element_blank())
```

> No clear changes in cell proportions are observed between the two conditions. Next we
investigate the PHD2cKO phenotype from the axis of Bnip3-Spp1 observed in the in-vitro
experment.

## PHD2cKO phenotype in myeloid cells

> We will map the in-vitro PHD2cKO signature onto myeloid cells as a proxy of PHD2cKO penetrance in-vivo. 
> Next we connect the PHD2cKO signature, which is derived from in-vitro cell culture, with
Bnip3 and Spp1 expression.

We first merge the two resident macrophage populations and discard two populations: 
the not assigned and proliferating populations. Note that we are going to explore the
PHD2KO-Bnip3-Spp1 axis in the main myeloid leukocyte groups.

```{r merge_myeloid}
# Drop odd clusters
M.i <- M.i[, !Idents(M.i) %in% c("n.a.", "Proliferating")]
# Merge Residen-Mac (2 clusters)
Idents(M.i) <- factor(gsub("(Resident-Mac)_[12]$",
			   "\\1", as.character(Idents(M.i))),
		      levels=unique(gsub("(Resident-Mac)_[12]$", 
					 "\\1", levels((M.i))))
		      )

# We save the sample size
write.table(as.data.frame(table(Group=Idents(M.i))),
	    file=paste0(DATADIR,"/nCell_per_group.csv"),
	    sep=",", row.names=FALSE, col.names=TRUE, quote=FALSE)

```

### PHD2cKO signature
We aim to characterize the impact of PHD2cKO in the in-vivo model. In theory, PHD2-KO is specific
of myeloid lineage. However the penetrance of the knock-out is not 100%. So far We would expect
higher PHD2-KO signal in PHD2cKO condition. To explore this, and potentially identify which cells
are undergoing PHD2-KO, we measure the expression of PHD2cKO signature derived from the in-vitro
bulk profiling of BMDM macrophages (PHD2cKO vs WT).
For this, we use AUCell method using top50 up-regulated genes from the signature.

```{r define_signature}
# Load in-vitro signature from the bulk analysis
invitro.dge <- read.table("../Analysis_invitro_bulk/output/01_bulk_dge/MC_PHD2_diffexpr.tsv", 
			  sep="\t", header=TRUE, stringsAsFactors = FALSE)
invitro.sign <- sort(setNames(invitro.dge$t,invitro.dge$genes),decreasing = TRUE)
# Get common genes
common <- intersect(rownames(S), names(invitro.sign))
# Discard certain genes that are driving
avoid_genes <- c("Bnip3", "Spp1")
common <- setdiff(common, avoid_genes)
# Define signature and show
invitro.gs <- names(sort(invitro.sign[common],decreasing = TRUE))[1:50]
print(invitro.gs)
# Export used signature as supplementary table
# NOTE: this is just a subset of the original table above (i.e. DGE fom PHD2cKO BMDM macrophages)
write.table(invitro.dge[invitro.dge$genes %in% invitro.gs, ], sep=",",
	    file=paste0(DATADIR, "/PHD2cKO_BMDM-derived_signature.csv"),
	    row.names=FALSE, col.names=TRUE, quote=FALSE)
```

```{r aucell_phd2ko_ML, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
# Build rankings for AUC
cells_rankings <- AUCell_buildRankings(as.matrix(M.i@assays$RNA@data))
# Use top N (e.g. 50) to calculate enrichment at single-cell level
cells_AUC <- AUCell_calcAUC(list("invitro"=invitro.gs), cells_rankings, aucMaxRank=nrow(cells_rankings)*0.05)
# Explore thresholds
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=FALSE, assign=TRUE) 
# Get AUCell scores  
AUC.sign <- getAUC(cells_AUC)[1,]
stopifnot(colnames(M.i) == names(AUC.sign))
M.i$PHD2cKO <- AUC.sign
rm(AUC.sign, cells_rankings, cells_AUC, cells_assignment)
```

We compare the signature expression between the two conditions using wilcox test. We also
show r effect sizes.

```{r wilcox_myeloid_PHD2cKO}
wPHD2cKO_stats <- t(sapply(levels(M.i), function(cell) {
				   cellIds <- WhichCells(M.i, idents=cell)
				   # The test
				   wilcox.test_stats(xy=M.i$PHD2cKO[cellIds],
						     gr=M.i$stim[cellIds])
			   }))
wPHD2cKO_stats <- as.data.frame(wPHD2cKO_stats)
wPHD2cKO_stats$adjpval <- p.adjust(wPHD2cKO_stats$pvalue, method="fdr")
wPHD2cKO_stats$significance <- tagSignif(wPHD2cKO_stats$adjpval)
print(wPHD2cKO_stats)
# Save it
write.table(wPHD2cKO_stats, sep=",",
	    file=paste0(DATADIR,"/PHD2cKO_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```

```{r vlnsplit_ML_PHD2cKO, fig.with=10, fig.height=6, warning=FALSE}
RES <- wPHD2cKO_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = M.i, gene=NULL,meta="PHD2cKO",
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.005, pt.alpha=0.4) +
ggtitle("PHD2cKO signature")
```

### Bnip3 expression

We perform the same comparison but for Bnip3 expression.

```{r wilcox_bnip3}
wBnip3_stats <- t(sapply(levels(M.i), function(cell) {
				   cellIds <- WhichCells(M.i, idents=cell)
				   # The test
				   wilcox.test_stats(xy=M.i@assays$RNA@data["Bnip3", cellIds],
						     gr=M.i$stim[cellIds])
			   }))
wBnip3_stats <- as.data.frame(wBnip3_stats)
wBnip3_stats$adjpval <- p.adjust(wBnip3_stats$pvalue, method="fdr")
wBnip3_stats$significance <- tagSignif(wBnip3_stats$adjpval)
print(wBnip3_stats)
# Save it
write.table(wBnip3_stats, sep=",",
	    file=paste0(DATADIR,"/Bnip3_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```


```{r vlnsplit_ML_Bnip3, fig.with=10, fig.height=6, warning=FALSE}
RES <- wBnip3_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = M.i, gene="Bnip3", meta=NULL,
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.1, pt.alpha=0.4)
```

### Spp1 expression

Similarly to Spp1 expression, we compare both conditions among myeloid leukocytes.

```{r wilcox_spp1}
wSpp1_stats <- t(sapply(levels(M.i), function(cell) {
				   cellIds <- WhichCells(M.i, idents=cell)
				   # The test
				   wilcox.test_stats(xy=M.i@assays$RNA@data["Spp1", cellIds],
						     gr=M.i$stim[cellIds])
			   }))
wSpp1_stats <- as.data.frame(wSpp1_stats)
wSpp1_stats$adjpval <- p.adjust(wSpp1_stats$pvalue, method="fdr")
wSpp1_stats$significance <- tagSignif(wSpp1_stats$adjpval)
print(wSpp1_stats)
# Save it
write.table(wSpp1_stats, sep=",",
	    file=paste0(DATADIR,"/Spp1_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```

```{r vlnsplit_ML_Spp1, fig.with=10, fig.height=6, warning=FALSE}
RES <- wSpp1_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = M.i, gene="Spp1", meta=NULL,
	stats=RES,
	vlnsplit = TRUE, fontTXT, nCell.y=-0.3, pt.alpha=0.4)
```

## Conclusion

> We are able to identify the major myeloid leukocytes from atherosclerotic plaques (Zernecke et al. 2020).
The PHD2cKO is well recovered in-vivo, where most of the myeloid leukocytes presents higher expression of 
this signature in PHD2cKO condition (medium-to-large effect sizes), with the exception of moDC and monocytes. 
The data suggests that PHD2cKO condition is associated with higher expression of Bnip3 in Macrophages, although 
the zero-inflation of single-cell data damps this statistical test. Trem2-foamy Macrophages are the main source
of Spp1 ligand expression in Atheroscleroic plaques, and PHD2cKO condition is associated to a higher 
expression in the distribution (FDR-adjusted p-value from wilcox test p<0.05, small effect size).

## Save the Seurat Object
```{r seurat_obj}
saveRDS(S, paste0(DATADIR,"/S.rds"));
saveRDS(M.i, paste0(DATADIR,"/M.rds"));
```

## SessionInfo
```{r}
sessionInfo()

{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```
