---
title: "Role of PHD2 atherogenesis in in-vivo plaque Macrophages"
author: "Javier Perales-PatÃ³n - javier.perales@bioquant.uni-heidelberg.de - ORCID: 0000-0003-0780-6683" 
output: github_document
---

> We have extensive data that supports the impact of PHD2 inhibition on promoting an aggravated 
atherogenesis phenotype via Macrophage apoptosis and its downstream signaling with Bnip3-Spp1 expression.
Previously we have shown that PHD2-KO is well recovered in most of the 
in-vivo myeloid leukocytes from the PHD2cKO condition. Bnip3, which is associated with the phenotype 
and macrophage apoptosis, seems to be up-regulated in said comparison in-vivo. Spp1, the prediced 
ligand released by Macrophages to modulate fibrotic fibroblasts based on the in-vitro data, seems to be
mainly released by TREM2hi Macrophages. Thus, there is concordance between in-vitro and in-vivo
data, but the (in-vivo) single-cell RNAseq data might provide higher resolution. Mainly pointing out 
TREM2hi macrophages as main players. Thus we investigate this line further here.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup
We define a random seed number for reproducibility, file structure for the output, 
and load essential libraries

### Environment
```{r env}
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./04_macrophage_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png','tiff'))
knitr::opts_chunk$set(dpi=300)
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);
```

### Load libraries
```{r}
suppressPackageStartupMessages(require(Seurat))
suppressPackageStartupMessages(require(GSEABase))
suppressPackageStartupMessages(require(cowplot))
suppressPackageStartupMessages(require(genesorteR))
suppressPackageStartupMessages(require(fgsea))
suppressPackageStartupMessages(require(ggplot2))
suppressPackageStartupMessages(require(ggrepel))
suppressPackageStartupMessages(require(AUCell))
suppressPackageStartupMessages(require(viper))
suppressPackageStartupMessages(require(purrr))
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(require(openxlsx))
suppressPackageStartupMessages(require(ComplexHeatmap))
suppressPackageStartupMessages(require(clusterProfiler))
suppressPackageStartupMessages(require(ggsignif))
source("../src/graphics.R")
source("../src/seurat_fx.R")
source("../src/wilcox_fx.R")
```

## Load data
Read the Seurat Object from third step (Myeloid leukocyte map).
```{r load_seurat}
# Input data
sobj <- "./03_myeloidlineage_output/data/M.rds"
# Read 
if(file.exists(sobj)) {
	M <- readRDS(sobj)
} else {
	stop("ERROR: Seurat object does not exist. Run 03.rmd to generate it.")
}
```

We focus on Macrophages in this analysis. So we get the cellIds for this population.
We show the sample size of each macrophage class for the records. It seems our data
presents low representation of IFNIC  macrophage population. Thus we might have to drop 
it from 2-group comparisons.

```{r macrophage_cellIds}
MacClasses <- grep("-Mac", levels(M), value=TRUE)
print(MacClasses)

MacIDs <- WhichCells(M, ident = MacClasses)
# Total cells
length(MacIDs)
# Sample size for each class
table("Class"=Idents(M[, MacIDs]))
# We save the sample size
write.table(as.data.frame(table("Class"=Idents(M[, MacIDs]))),
	    file=paste0(DATADIR,"/nCell_per_class.csv"),
	    sep=",", row.names=FALSE, col.names=TRUE, quote=FALSE)
```

We confirm the identity of the Macrophages using the markers from Zernecke 2020 again.
```{r heatmap_Zernecke_macrophages, fig.width=10, fig.height=10, dpi=300, warning=FALSE}
# Top10 markers for each Myeloid Leukocytes found in Meta-analysis review Zernecke et al. 2020
ML.markers <- getGmt("../data/markers/leukocytes_Zernecke2020.gmt")
ML.markers <- ML.markers[grep("_Mac", names(ML.markers))]

hp <- DoHeatmap3(SeuratObject=M[, MacIDs], GSC=ML.markers, assay="RNA", res="Idents", 
	   row_names_size=10, column_title_size=0,
	   show_hr=FALSE, legend_nrow=2, fontfamily=fontTXT) 
draw(hp, heatmap_legend_side="right", annotation_legend_side="bottom")
```

We confirm the TREM2hi macrophages identity looking into detail in the all differentially
expressed genes from Zernecke et al 2020.

```{r heatmap_Zernecke_TREM2hi, fig.width=14, fig.height=10, dpi=300, warning=FALSE}
trem2 <- read.table("../data/markers/Trem2_foamy-Mac_DEGs_Zernecke2020.tsv", 
		     sep="\t", header=TRUE, stringsAsFactors=FALSE)

trem2 <- trem2[trem2$p_val_adj < 0.05, ]

trem2 <- GeneSetCollection(GeneSet(trem2$gene, setName="Trem2.Mac"))
hp <- DoHeatmap3(SeuratObject=M[, MacIDs], GSC=trem2, assay="RNA", res="Idents", 
	   row_names_size=0, column_title_size=0,
	   show_hr=FALSE, legend_nrow=2, fontfamily=fontTXT) 
draw(hp, heatmap_legend_side="right", annotation_legend_side="bottom")
```


## Comparison between PHD2 functional perturbation

>The in-vivo model carries a PHD2 knock-out in the myeloid lineage with an effectivity of ~30% (estimated)
of the cells. We have shown that PHD2cKO condition presents higher penetrance of BMDM-derived PHD2cKO signature.
The section will investigate if the penetrance of this gene perturbation is enough to be captured in the 
average population of the pooled PHD2cKO mice.

### Differential gene expression
We perform a standard differential expression with Bonferroni correction.

The scatter plot between the average expression of the two populations highlight Spp1
as one of the most differentially expressed genes in PHD2cKO condition in Macrophages.

```{r scatterplot_macrophages_PHD2cKO_vs_PHD2wt, fig.width=5, fig.height=5, dpi=300}
S1 <- M[, MacIDs] # Select macrophages to pool for a test diff expr

Idents(S1) <- "stim"
avg <- log1p(AverageExpression(S1, verbose = FALSE)$RNA)
  
rnk <- setNames(avg[,"PHD2cKO"] - avg[,"WT"], rownames(avg))
top <- names(sort(abs(rnk),decreasing = TRUE)[1:15])
  
ggplot(avg, aes(WT, PHD2cKO)) + geom_point(alpha=0.6) + ggtitle(label = "Macrophages") + 
	xlab("WT") + ylab("PHD2cKO")+
	geom_label_repel(data=avg[top,], aes(label=top), size=5) + 
	theme_cowplot() +
      theme(axis.title = element_text(size=22),
            title = element_text(size=22, hjust = 0.5),
            axis.text = element_text(size=18),
            axis.title.x = element_text(size=18),
            axis.title.y = element_text(size=18))
    
  
dge <- FindMarkers(object = S1, ident.1 = "PHD2cKO", ident.2 = "WT", 
		  logfc.threshold=0, min.pct=0)
cols_names <- colnames(dge)
# Add two extra cols
dge$cluster <- "Macrophage PHD2cKO vs PHD2wt"
dge$gene <- rownames(dge)

dge <- dge[,c("cluster", "gene", cols_names)]
   
write.table(dge,
            file = paste0(DATADIR,"/Macrophages_avg_PHD2cKO_vs_PHD2wt_DEGs.tsv"),
            sep="\t",col.names = TRUE, row.names = FALSE, quote=FALSE
)
```

We show the number of differentially expressed genes (Bonferroni-adjusted pvalue < 0.05):
```{r}
table(sign(dge$avg_logFC) * as.integer(dge$p_val_adj < 0.05))
```

> We could conclude that there are not large differences in the average gene expression of 
macrophages based on condition (PHD2 KO). 

Next efforts will be made to dissect the main
transcriptional programme of this perturbation (Hypoxia) and explore the Bnip3-Spp1 axis
in the different classes of macrophages.

## Dissection of hypoxia response and PHD2 functional impairment

> We know that PHD2 is an important regulator in hypoxia response. We perform a
transcriptional dissection of hypoxia driven by the hypothesis that 
PHD2-KO perturbation triggers a strong hypoxia response. Thus it is expected that the 
PHD2-cKO condition presents an over-representation of cells responding to hypoxia.

### Dorothea focused on Hif1a
We calculate Hif1a transcription factor activities using dorothea. We know that Hif1a is 
the canonical transcription factor for hypoxia response, and PHD2 interacts with it to trigger
such a response.

```{r vln_macrophages_dorothea_hif1a, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
# load regulons
df2regulon <- function(df, regulator_name="tf") {
  regulon = df %>% split(.[regulator_name]) %>% map(function(dat) {
    targets = setNames(dat$mor, dat$target)
    likelihood = dat$likelihood
    list(tfmode = targets, likelihood = likelihood)
  })
  return(regulon)
}

regulon.df <- read.table("../data/Prior/dorothea_regulon_mouse_v1.csv", sep=",", 
			 header=TRUE, stringsAsFactors = FALSE)
regul <- df2regulon(df=regulon.df)

# Calculate TF activities
TF <- viper(eset = as.matrix(M@assays$RNA@data), regulon = regul,
              nes = T, minsize = 4,
              eset.filter = F, adaptive.size = F,
	      verbose=FALSE)
  
# Add them as metadata
stopifnot(colnames(M) == colnames(TF))
M$Hif1a_activity <- TF["Hif1a",]
rm(TF)

# Visualization focused on Macrophages population
VlnPlot.stim(M[, MacIDs], # Select Macrophages
     meta.feature = "Hif1a_activity", ylab="Hif1a activity", 
     fontTXT=fontTXT)

```

### PROGENy focused on hypoxia response
We calculate a score of hypoxia response using progeny as well.

```{r vln_progeny_hypoxia_macrophages, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
### Progeny ####
progeny.mat <- read.table("../data/Prior/progeny_matrix_mouse_v1.txt",sep=",",header=TRUE)
rownames(progeny.mat) <- progeny.mat$X
progeny.mat <- progeny.mat[which(colnames(progeny.mat)!="X")]
progeny.mat <- as.matrix(progeny.mat)

common <- intersect(rownames(M), rownames(progeny.mat))
  
prog <- t(as.matrix(M@assays$RNA@data[common,])) %*% progeny.mat[common,]
rn <- rownames(prog)
prog <- apply(prog,2,scale)
rownames(prog) <- rn
prog <- t(prog)
  
stopifnot(colnames(M) == colnames(prog))
M$Hypoxia_response <- prog["Hypoxia",]
rm(common,prog)

# Visualization focused on Macrophages population
VlnPlot.stim(M[, MacIDs], # Macrophages
     meta.feature = "Hypoxia_response", ylab="Hypoxia response (Score)", 
     fontTXT=fontTXT)
```

### Predisposition to hypoxia response

To understand better how macrophages response to hypoxia in atherosclerotic plaques,
we check whether there is any macrophage subset with higher response in the WT 
condition (predisposition to hypoxia).

```{r vlnsplit_hypoxia_predisposition_WTmac, fig.width=5, fig.height=7, dpi=300, warning=FALSE}
S3 <- M[, MacIDs]
WT_Mac <- S3[, S3$stim=="WT"]

wHypoxiaTrem2_stats <- t(sapply(setdiff(levels(WT_Mac), "TREM2hi-Mac"), function(cell) {
				   cellIds <- WhichCells(WT_Mac, idents=c(cell, "TREM2hi-Mac"))
				   # The test
				   wilcox.test_stats(xy=WT_Mac[, cellIds]@meta.data$Hypoxia_response,
						     gr=factor(as.character(Idents(WT_Mac)[cellIds])))

				   
					 }))
wHypoxiaTrem2_stats <- as.data.frame(wHypoxiaTrem2_stats)
wHypoxiaTrem2_stats$adjpval <- p.adjust(wHypoxiaTrem2_stats$pvalue, method="fdr")
wHypoxiaTrem2_stats$significance <- tagSignif(wHypoxiaTrem2_stats$adjpval)

RES <- wHypoxiaTrem2_stats
RES$r <- format(round(RES$r, 2), nsmall=2)
RES$ann <- paste("r=",RES$r,"\n",RES$significance)
RES <- RES[rev(rownames(RES)), ]

gg1 <- plotVln(SeuratObject = WT_Mac, gene=NULL, meta="Hypoxia_response",
	stats=NULL,
	vlnsplit = FALSE, fontTXT,  nCell.y=-3.2, pt.alpha=0.4) +
	ylab("Scaled activity score") + NoLegend() +
	coord_cartesian(ylim = c(-3.5, 7), clip="off")

gg1 + geom_signif(comparisons = sapply(rownames(RES), function(z) c(z, "TREM2hi-Mac"), simplify=FALSE),
		  colour="black",
	    y_position = seq(from=2.5, by=1.2, length.out=4),
		  annotations= RES$ann)
```


### Enrichment of in-vitro PHD2-KO signature
In the previous step, [03_myeloidlineage.md](03_myeloidlineage.md), we have already
calculated a single-cell score of the enrichment of in-vitro PHD2-KO signature among
myeloid leukocytes, including all groups of macrophages.

> NOTE: Since we have already statistically tested this hypothesis among myeloid leukocytes,
we use the statistics from previous analysis so we do not have to correct for multiple
testing twice.

Here we show it again, but restricted to Macrophages.

```{r vlnsplit_macrophage_PHD2cKO, fig.width=7, fig.height=6, warning=FALSE}
wPHD2cKO_stats <- read.table(file=paste0("./03_myeloidlineage_output/data",
					 "/PHD2cKO_wilcox_stats.csv"),
			     sep=",", header=TRUE, stringsAsFactors=FALSE)
rownames(wPHD2cKO_stats) <- wPHD2cKO_stats$X

RES <- wPHD2cKO_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

stopifnot(all(levels(M)==rownames(RES)))
plotVln(SeuratObject = M[, grep("-Mac", Idents(M))], 
	gene=NULL,meta="PHD2cKO",
	stats=RES[grep("-Mac", rownames(RES)), ],
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.005, pt.alpha=0.4) +
ggtitle("PHD2cKO signature")
```


### Pair-wise comparison of these markers of hypoxia response

```{r pairs_hypoxia_macrophages, fig.width=9, fig.height=9, dpi=300, warning=FALSE}
# Define colors for ggplot
cond.cols <- c("WT"="grey", "PHD2cKO"="red")

panel.cor <- function(x, y, digits=2, prefix="", cex.cor, family=fontTXT, ...)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y, method="pearson"))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  txt <- paste(prefix, txt, sep="")
  if(missing(cex.cor)) cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r, family=family)
}

# Visualization
pairs(M@meta.data[MacIDs,
      		c("Hif1a_activity","Hypoxia_response","PHD2cKO")],
      lower.panel=panel.smooth, upper.panel=panel.cor, cex.labels = 2.6, cex.axis=2.3, las=1,
      family=fontTXT,
      pch=21,
      bg=cond.cols[M@meta.data[MacIDs, "stim"]])
```

> Conclusion: We could conclude that PHD2-cKO condition presents a skewed distribution of cells towards 
hypoxia response as compared to PHD2-wt condition. The PHD2cKO signature expression is highly correlated
to Hif1a transcription factor and hypoxia response activities. Their activity and expression is higher in
macrophages from PHD2cKO condition. As a baseline, wild-type TREM2hi macrophages seems to present 
a slight higher response to hypoxia in atherosclerotic plaques as compared to other subsets such as 
resident and inflammatory macrophages.

## Investigation of TREM2hi macrophages

We look into the comparison between PHD2cKO and WT TREM2hi macrophages. 
We observed that there is a tendency of Spp1 over-expression in PHD2cKO macrophages.
Actually Spp1 has been described as a marker gene of TREM2hi macrophages (Zernecke et al 2020).
Thus we look into detail for over-expression of other TREM2hi macrophages beyond Spp1
in all subsets.

First we look at the top 10 over-expressed genes in TREM2hi.
```{r vlnsplit_TREM2hi_Signature_top10, fig.width=16, fig.height=8, warning=FALSE}
plots <- sapply(unlist(geneIds(trem2))[1:10], function(trem2_gene) {
	       plotVln(SeuratObject = M[, MacIDs], gene=trem2_gene, meta=NULL,
		       stats=NULL,
		       vlnsplit = TRUE, fontTXT,  nCell.y=-0.5, pt.alpha=0.4) +
       		NoLegend()
	}, simplify=FALSE)

plot_grid(plotlist = plots, nrow=2)
```

Second we estimate a single-cell enrichment score of the over-expression of TREM2hi signature to
see if there is any difference in PHD2cKO condition among macrophage subsets.

```{r aucell_TREM2hi_signature, fig.width=5, fig.height=7, dpi=300, warning=FALSE}
# Build rankings for AUC
cells_rankings <- AUCell_buildRankings(as.matrix(M[, MacIDs]@assays$RNA@data))
# Use top N (e.g. 50) to calculate enrichment at single-cell level
cells_AUC <- AUCell_calcAUC(lapply(geneIds(trem2), function(gs) head(gs,50)), 
			cells_rankings, 
			 aucMaxRank=nrow(cells_rankings)*0.05)
# Explore thresholds
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=FALSE, assign=TRUE) 
# Get AUCell scores  
AUC.sign <- getAUC(cells_AUC)[1,]
stopifnot(colnames(M[, MacIDs]) == names(AUC.sign))
M$trem2.Mac <- NA
M$trem2.Mac[MacIDs] <- AUC.sign
rm(AUC.sign, cells_rankings, cells_AUC, cells_assignment)
```

```{r vlnsplit_TREM2hi_SignatureMapping_macrophages, fig.width=6, fig.height=7, warning=FALSE}
wTrem2_stats <- t(sapply(levels(M[, MacIDs]), function(cell) {
				   cellIds <- WhichCells(M[, MacIDs], idents=cell)
				   # The test
				   wilcox.test_stats(xy=M[, cellIds]@meta.data$trem2.Mac,
						     gr=M[, cellIds]$stim)
			   }))
wTrem2_stats <- as.data.frame(wTrem2_stats)
wTrem2_stats$adjpval <- p.adjust(wTrem2_stats$pvalue, method="fdr")
wTrem2_stats$significance <- tagSignif(wTrem2_stats$adjpval)
print(wTrem2_stats)

RES <- wTrem2_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = M[, MacIDs], gene=NULL, meta="trem2.Mac",
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=0, pt.alpha=0.4) +
	ylab("Enrichment score (AUC)") + ggtitle("Trem2-foamy Macrophage signature") +
      guides(fill= guide_legend(nrow=2, byrow=TRUE)) 
```

> Conclusion: We have correctly identified TREM2hi macrophages. These over-express Spp1, which is 
up-regulated upon PHD2cKO perturbation. However, it seems that there is no difference in the
expression of the TREM2hi signature (i.e. the rest of the over-expressed genes in TREM2hi macrophages 
as compared to other myeloid leukocytes subset) upon PHD2cKO, but mainly Spp1.



## PHD2KO-Bnip3-Spp1 axis

We already have shown that PHD2cKO signature is recovered in PHD2cKO condition. 
What about the Bnip3 and Spp1 expression? These has been already explored among
other myeloid leukocytes, but it is shown here focused on macrophages. 

### Bnip3 expression across macrophages

> NOTE: Since we have already statistically tested this hypothesis among myeloid leukocytes,
we use the statistics from previous analysis so we do not have to correct for multiple
testing twice.

```{r vlnsplit_macrophage_Bnip3, fig.width=7, fig.height=6, warning=FALSE}
wBnip3_stats <- read.table(file=paste0("./03_myeloidlineage_output/data",
					 "/Bnip3_wilcox_stats.csv"),
			     sep=",", header=TRUE, stringsAsFactors=FALSE)
rownames(wBnip3_stats) <- wBnip3_stats$X

RES <- wBnip3_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

stopifnot(all(levels(M)==rownames(RES)))
vln_Bnip3 <- plotVln(SeuratObject = M[, grep("-Mac", Idents(M))], 
		     gene="Bnip3",meta=NULL,
		     stats=RES[grep("-Mac", rownames(RES)), ],
		     vlnsplit = TRUE, fontTXT,  nCell.y=-0.3, pt.alpha=0.4) 
vln_Bnip3
```


### Spp1 expression across macrophages

```{r vlnsplit_macrophage_Spp1, fig.width=7, fig.height=6, warning=FALSE}
wSpp1_stats <- read.table(file=paste0("./03_myeloidlineage_output/data",
					 "/Spp1_wilcox_stats.csv"),
			     sep=",", header=TRUE, stringsAsFactors=FALSE)
rownames(wSpp1_stats) <- wSpp1_stats$X

RES <- wSpp1_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

stopifnot(all(levels(M)==rownames(RES)))
vln_Spp1 <- plotVln(SeuratObject = M[, grep("-Mac", Idents(M))], 
		    gene="Spp1",meta=NULL,
		    stats=RES[grep("-Mac", rownames(RES)), ],
		    vlnsplit = TRUE, fontTXT,  nCell.y=-0.3, pt.alpha=0.4) 
vln_Spp1
```

> Conclusion: Bnip3 is hampered by drop-outs in single-cell profiling (zero-inflation), 
thus it makes difficult to compare groups. However, it seems that Bnip3 is up-regulated 
upon PHD2cKO. In contrast, PHD2cKO-induced up-regulation of Spp1 expression seems to be 
specific of TREM2hi macrophages.

Next we stratify macrophages to compare those undergoing complete PHD2-KO.

### Stratification of cells by BMDM-derived PHD2cKO signature
We just observed that the penetrance of PHD2-KO is not complete in the PHD2cKO condition.
Thus we use the expression of the in-vitro signature to compare those cells undergoing
PHD2 versus WT. For this, we choose an arbitrary threshold, the 3rd quartile of the signature
expression in PHD2cKO condition. We select these cells to be compared to WT macrophages that are 
not responding to Hypoxia/PHD2cKO, by using the threshold of being below the Q3 cutoff. We opt
to discard WT \>Q3 signature expression to avoid downstream signaling of hypoxia response, but
actually just only few WT macrophages (n=7) are dropped with this criteria, while 419 WT cells 
(low) are retrieved for the comparison. The final comparison includes 79 PHD2cKO cells vs 
419 WT cells.

```{r}
cutoff <- quantile(M@meta.data[colnames(M) %in% MacIDs & M$stim=="PHD2cKO", "PHD2cKO"], 
		   probs=0.75) 

M$PHD2cKO_class <- NA

M$PHD2cKO_class[MacIDs] <- ifelse(M$PHD2cKO[MacIDs] > cutoff,
						   "High", "Low")
 
#NOTE: here it is where we concatenate condition + stratified class by PHD2cKO signature
# so later we could just compare the two groups of interest:
# - PHD2cKO macrophages above Q3 signature expression (high)
# - WT macrophages below Q3 signature expression (Low)
M$PHD2cKO_class2 <- paste0(M$stim,"_",M$PHD2cKO_class)

# Show cutoff used
print(cutoff)
# Save sample sizes
table(Strata=M$PHD2cKO_class2[MacIDs])
# We save the sample size
write.table(as.data.frame(table("Strata"=Idents(M[, MacIDs]))),
	    file=paste0(DATADIR,"/nCell_per_class.csv"),
	    sep=",", row.names=FALSE, col.names=TRUE, quote=FALSE)
```

The cutoff of Q3 in PHD2cKO is `r cutoff`.

For the records, we just show which cells has been selected below/above that line.

```{r vln_aucell_phd2ko_macrophages_wCutoff, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
# Visualization focused on Macrophages population
VlnPlot.stim(M[,MacIDs], 
     meta.feature = "PHD2cKO", ylab="Expression of PHD2cKO signature (AUC)", 
     fontTXT=fontTXT) + geom_hline(yintercept=cutoff)
```

We retrieve only the macrophages involved in the group of interest from now on. So the rest
of the analysis is focused on this stratification of cells

```{r sel_mac}
MacIDs <- MacIDs[which(M$PHD2cKO_class2[MacIDs] %in% c("PHD2cKO_High", "WT_Low"))]
```

Perform a genome-wide differential gene expression between the two stratified groups.

```{r dge_PHD2cKO_High}
S1 <- M[, MacIDs] 
Idents(S1) <- "PHD2cKO_class2"
dge <- FindMarkers(S1, ident.1 = "PHD2cKO_High", ident.2 = "WT_Low",
		   min.pct = 0, logfc.threshold = 0)

saveRDS(dge, file="Mac_PHD2cKO_High_DGE.Rds")
```

We generate a volcano plot to visually inspect the transcriptome-wide differences.
```{r volcano_PHD2cKOsign_high_vs_low, dpi=300, fig.width=8, fig.height=6, warning=FALSE}
Mac_interesting_genes <- getGmt("../data/Prior/MC_PHD2_diff.gmt")
Mac_interesting_genes <- unlist(geneIds(Mac_interesting_genes))
if(any(!Mac_interesting_genes %in% rownames(S1))) {
	cat(paste("The following genes are not present in the gene expr matrix:","\n",
		  paste(setdiff(Mac_interesting_genes, rownames(S1)), collapse=","),
		  "\n"),
		  file=stdout())
	Mac_interesting_genes <- intersect(Mac_interesting_genes, rownames(S1))
}

dge$significant <- ifelse(dge$p_val_adj < 0.05,"Adj.P-value<0.05","n.s.")
dge$genes <- rownames(dge)
dge$show <- dge$genes %in% Mac_interesting_genes
 
p = ggplot(dge, aes(avg_logFC, -log10(p_val+0.001))) +
    geom_point(aes(col=significant), alpha=0.6) +
    scale_color_manual(values=c("red", "black"))
  
  
p+geom_label_repel(data=dge[dge$show,],
                          aes(label=genes), family=fontTXT, size=8,
                          force=2,
			  xlim = c(2.5,3.0),
                          nudge_y=0.05, direction = "y", segment.size = 0.5) + 
    coord_cartesian(xlim = c(-1.3, 2.3), clip = "off") +
    theme_cowplot() +
    xlab("avg log2FC") + ylab("-log10(pvalue + 0.001)") +
    theme(text = element_text(family=fontTXT, size=20),
	  legend.text = element_text(family=fontTXT, size=20),
	  axis.text = element_text(family=fontTXT, size=20),
    	  plot.margin = unit(c(0.3, 2.6, 0.1, 0.1), "cm"),
	  legend.position = "bottom",
	  legend.direction = "horizontal"
	  )
```

We observe two points here:

- Bnip3 is significantly over-expressed in High PHD2cKO signature macrophages.
- Spp1 follows similar trend, but it is not significant. It seems that there are
  some outliers over-expressing the gene, but not all macrophages. This makes sense 
  since it seems Spp1 over-expression is specific of only one class of macrophages:
  trem2-foamy class.


## Re-visit Bnip3 and Spp1 expression in stratified macrophages
> We just distilled which macrophages are undergoing PHD2-KO in the PHD2cKO condition
and which ones are WT sampled from WT condition. We compared their differential expression.
Now we re-visit a potential differential expression of Bnip3 and Spp1 following the evidences
from in-vitro experiments and data, but following a refined analysis with only these groups of 
macrophages (high and low PHD2cKO signature).

```{r dotplot_Bnip3Spp1, fig.width=7, fig.height=4}
S1 <- M[, MacIDs]
S1$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S1$PHD2cKO_class2])
S1$stim <- relevel(S1$stim, ref=grep("WT", levels(S1$stim), value=TRUE))
cnt <- table(S1$stim, Idents(S1))>10
sel2 <- colnames(cnt)[colSums(cnt)==2]

DotPlot(S1[, WhichCells(S1, ident=sel2)], 
	features=c("Bnip3", "Spp1"), 
	split.by="stim", cols=c("blue","red"))
```

### Bnip3 in high PHD2cKO Macrophages

```{r wilcox_bnip3_MacPHD2sign}
S1 <- M[, MacIDs]
S1$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S1$PHD2cKO_class2])
S1$stim <- relevel(S1$stim, ref=grep("WT", levels(S1$stim), value=TRUE))

wBnip3_stats <- t(sapply(levels(S1), function(cell) {
				   cellIds <- WhichCells(S1, idents=cell)
				   # The test
				   wilcox.test_stats(xy=S1@assays$RNA@data["Bnip3", cellIds],
						     gr=S1$stim[cellIds])
			   }))
wBnip3_stats <- as.data.frame(wBnip3_stats)
wBnip3_stats$adjpval <- p.adjust(wBnip3_stats$pvalue, method="fdr")
wBnip3_stats$significance <- tagSignif(wBnip3_stats$adjpval)
print(wBnip3_stats)
# Save it
write.table(wBnip3_stats, sep=",",
	    file=paste0(DATADIR,"/PHD2sign_Bnip3_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```

```{r vlnsplit_Bnip3_MacPHD2sign, fig.width=4, fig.height=6, warning=FALSE}
RES <- wBnip3_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = S1, gene="Bnip3", meta=NULL,
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.1, pt.alpha=0.4) +
      guides(fill= guide_legend(nrow=2, byrow = TRUE)) 
```


### Spp1 in high PHD2cKO Macrophages

```{r wilcox_spp1_MacPHD2sign}
S1 <- M[, MacIDs]
S1$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S1$PHD2cKO_class2])
S1$stim <- relevel(S1$stim, ref=grep("WT", levels(S1$stim), value=TRUE))

wSpp1_stats <- t(sapply(levels(S1), function(cell) {
				   cellIds <- WhichCells(S1, idents=cell)
				   # The test
				   wilcox.test_stats(xy=S1@assays$RNA@data["Spp1", cellIds],
						     gr=S1$stim[cellIds])
			   }))
wSpp1_stats <- as.data.frame(wSpp1_stats)
wSpp1_stats$adjpval <- p.adjust(wSpp1_stats$pvalue, method="fdr")
wSpp1_stats$significance <- tagSignif(wSpp1_stats$adjpval)
print(wSpp1_stats)
# Save it
write.table(wSpp1_stats, sep=",",
	    file=paste0(DATADIR,"/PHD2sign_Spp1_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```

```{r vlnsplit_Spp1_MacPHD2sign, fig.width= 4, fig.height=6, warning=FALSE}
RES <- wSpp1_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = S1, gene="Spp1", meta=NULL,
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.3, pt.alpha=0.4) +
      guides(fill= guide_legend(nrow=2, byrow=TRUE)) 
```

## Differences in TREM2hi macrophages
We investigate if any difference between TREM2hi macrophage.
We cannot describe cell numbers between the two conditions because of the experimental design where
a pool of mice were profiled at once per condition, but these proportions are described in the previous
analysis of the myeloid lineage. It seems proportions does not change between conditions.

Next we perform differential expression to uncover the main biological processes underlying PHD2cKO. For this,
we use the already stratification by low and high expression of the PHD2cKO signature derived from in-vitro.

```{r TREM2hi_dge}
TREM2hi.idx <- WhichCells(M, idents="TREM2hi-Mac")
S2 <- M[, TREM2hi.idx] 
Idents(S2) <- "PHD2cKO_class2"
dge <- FindMarkers(S2, ident.1 = "PHD2cKO_High", ident.2 = "WT_Low",
		   min.pct = 0, logfc.threshold = 0)

saveRDS(dge, file=paste0(DATADIR,"/TREM2hi-Mac_PHD2cKO_High_DGE.Rds"))

write.table(as.data.frame(dge), 
	    file=paste0(DATADIR,"/TREM2hi-Mac_PHD2cKO_High_DGE.csv"), 
	    sep=",", row.names=TRUE, 
	    col.names=TRUE, quote=FALSE)
```

```{r TREM2hi_dge_volcano, dpi=300, fig.width=8, fig.height=6, warning=FALSE}
dge$significant <- ifelse(dge$p_val_adj < 0.05,"Adj.P-value<0.05","n.s.")
dge$genes <- rownames(dge)
dge$show <- ifelse(dge$p_val_adj < 0.05, TRUE, FALSE)
 
p = ggplot(dge, aes(avg_logFC, -log10(p_val+0.001))) +
    geom_point(aes(col=significant), alpha=0.6) +
    scale_color_manual(values=c("red", "black"))
  
  
p+geom_label_repel(data=dge[dge$show,],
                          aes(label=genes), family=fontTXT, size=8,
                          force=1,
# 			  xlim = c(2.5,3.0),
 			  ylim = c(3.5,5),
                          nudge_y=0.5, direction = "y") + 
    coord_cartesian(ylim = c(0, 6), clip = "off") +
    theme_cowplot() +
    xlab("avg log2FC") + ylab("-log10(pvalue + 0.001)") +
    theme(text = element_text(family=fontTXT, size=20),
	  legend.text = element_text(family=fontTXT, size=20),
	  axis.text = element_text(family=fontTXT, size=20),
    	  plot.margin = unit(c(0.3, 2.6, 0.1, 0.1), "cm"),
	  legend.position = "bottom",
	  legend.direction = "horizontal"
	  )

```


We focus on hypoxia response. 
```{r TREM2hi_hypoxia}
TREM2hi.idx <- WhichCells(M, idents="TREM2hi-Mac")
S2 <- M[, TREM2hi.idx] 
S2$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S2$PHD2cKO_class2])
S2$stim <- relevel(S2$stim, ref=grep("WT", levels(S2$stim), value=TRUE))
S2 <- S2[,!is.na(S2$stim)]

wHypoxia_stats <- t(sapply(levels(S2), function(cell) {
				   cellIds <- WhichCells(S2, idents=cell)
				   # The test
				   wilcox.test_stats(xy=S2@meta.data$Hypoxia_response,
						     gr=S2$stim)
			   }))
wHypoxia_stats <- as.data.frame(wHypoxia_stats)
wHypoxia_stats$adjpval <- p.adjust(wHypoxia_stats$pvalue, method="fdr")
wHypoxia_stats$significance <- tagSignif(wHypoxia_stats$adjpval)
print(wHypoxia_stats)
#Save it
write.table(wHypoxia_stats, sep=",",
 	    file=paste0(DATADIR,"/TREM2hi_PHD2sign_Hypoxia_wilcox_stats.csv"),
 	    row.names=TRUE, col.names=NA)
```

We observe a high difference on this pathway using PROGENy. See violin plot below with 
statistics.

```{r vlnsplit_Hypoxia_TREM2hiMacPHD2sign, fig.width= 4, fig.height=6, warning=FALSE}
RES <- wHypoxia_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = S2, gene=NULL, meta="Hypoxia_response",
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-2.7, pt.alpha=0.4) +
	ylab("Scaled activity score") +
      guides(fill= guide_legend(nrow=2, byrow=TRUE)) 
```


### Over-representation analysis
Finally we conclude our analysis with an over-representation analysis of differentially expressed genes.

```{r ORA_fx_aux}
# Define auxiliary functions for ORA
prepareGSC <- function(gmt_file, mmu2hsa) {
 	GSC <- getGmt(gmt_file)
 	GSC <- lapply(geneIds(GSC), function(z) unlist(sapply(z, function(j) mmu2hsa[[j]])))
	GSC_t2g <- data.frame(term=unlist(sapply(names(GSC), function(k) rep(k, length(GSC[[k]])))),
 		     gene = unlist(GSC))

	return(GSC_t2g)
}

enricher_saveExcel <- function(dge,res, fl) {
	wb <- createWorkbook()
	# TopTags
	addWorksheet(wb, sheetName = "DiffExpr")
	writeData(wb, sheet ="DiffExpr", x=as.data.frame(dge), rowNames=TRUE)
	for(idx in names(res)) {
		# GO terms
		addWorksheet(wb, sheetName = idx)
		writeData(wb, sheet =idx, x=res[[idx]]@result, 
			  rowNames=FALSE)
	}
	saveWorkbook(wb, file =  fl, overwrite = TRUE)
}

enricher_saveCSV <- function(res, fl) {
	for(idx in names(res)) {
		fl2 <- gsub("ORA",paste0("ORA", "_", idx), fl)
		# GO terms
		write.table(as.data.frame(res[[idx]]@result), file=fl2, sep=",", 
		    row.names=FALSE, col.names=TRUE, quote=FALSE)
	}
}
```

```{r ORA_libs}
# Lib of hsa2mmu orthologs
mmu2hsa <- readRDS("../data/Gene_annotation/hgnc2mgi_hsa2mmu.rds")

# Prepare GSC as data.frames for enrichr()
GO <- prepareGSC("../data/MSigDB/c5.bp.v7.0.symbols.gmt", mmu2hsa)
H <- prepareGSC("../data/MSigDB/h.all.v6.2.symbols.gmt", mmu2hsa)
```

```{r ORA_test}
# Get differentially expressed genes for ORA
TREM2hi_dge <- rownames(dge)[dge$p_val_adj < 0.05]
# TREM2hi_dge <- head(rownames(dge), 100)
# Show summary
cat("[INFO] selected genes for ORA:\n",file=stdout())
cat("[INFO] \t Among which are differentially expressed genes (BH-adj pval <0.05):\n",file=stdout())
print(table(dge[TREM2hi_dge, "p_val_adj"] < 0.05))
cat("[INFO] \t Distribution of log2-fold-changes in selection:\n",file=stdout())
print(summary(dge[TREM2hi_dge, "avg_logFC"]))
cat("[INFO] \t Distribution of absolute log2-fold-changes in selection:\n",file=stdout())
print(summary(abs(dge[TREM2hi_dge, "avg_logFC"])))

set.seed(1234)
GO.res <- enricher(TREM2hi_dge, TERM2GENE=GO, minGSSize = 10, maxGSSize = 200, pvalueCutoff = 0.05)
set.seed(1234)
H.res <- enricher(TREM2hi_dge, TERM2GENE=H, minGSSize = 10, maxGSSize = 200, pvalueCutoff = 0.05)
```

```{r ORA_bp_TREM2hi_PHD2cKOHigh_GOterms, fig.width=17, fig.height=10.5, dpi=300, warning=FALSE}
res <- GO.res@result
res$ID <- factor(res$ID, levels=rev(res$ID))
res$Ann <- paste0(res$GeneRatio,": ",res$geneID)

ggplot(head(res, 20), aes(y=-log10(p.adjust), x=ID)) + 
 	geom_bar(stat="identity", width=0.7, fill="grey") +
# 	geom_point(size=7) +
# 	geom_segment(aes(xend=ID), yend=0) + 
# 	expand_limits(y=0) + 
	geom_text(mapping=aes(y=0.01, x=ID, label=Ann), hjust=0, size=5.7) + 
	theme_cowplot() +
 	geom_hline(yintercept = -log10(0.05), lty=2, alpha=0.5) + 
 	coord_flip() +
	ylab("-log10(adjusted P-value)") +
	theme(legend.position="bottom",
	      legend.key.width=unit(1.5,"cm"),
	      legend.title=element_text(family=fontTXT, size=16),
	      legend.text=element_text(family=fontTXT, size=16),
	      axis.title.y=element_blank(),
	      axis.text.y=element_text(family=fontTXT, size=19),
	      axis.text.x=element_text(family=fontTXT, size=20),
	      axis.title.x=element_text(family=fontTXT, size=18))
```


```{r ORA_export}
enricher_saveExcel(dge, res=list("GO"=GO.res, "Hallmarks"=H.res),
		   fl = paste0(DATADIR,"/TREM2hi-Mac_PHD2cKO_High_ORA.xlsx"))

enricher_saveCSV(res=list("GO"=GO.res, "Hallmarks"=H.res),
		   fl = paste0(DATADIR,"/TREM2hi-Mac_PHD2cKO_High_ORA.csv"))
```




## Conclusion

> We have dissected PHD2cKO perturbation in all 5 classes of Macrophages previously described in
atherosclerotic plaques (Zernecke et al. 2020). The PHD2cKO condition presents uncomplete but 
enriched penetrance of the perturbation, and we were able to recover which cells are undergoing 
the perturbation by using a BMDM-derived PHD2cKO signature from in-vitro experiments. Doing so,
we were able to compare PHD2cKO and WT macrophages in the two conditions, and stratify them 
by those most responding to the perturbation. We confirmed that PHD2cKO leads to hypoxia response.
We also observed that Bnip3 is up-regulated among all macrophage classes (except for IFNIC,
for which we do not have enough observations). In opposite, Spp1 over-expression upon PHD2cKO is
specific of TREM2hi macrophages, suggesting that these are the source of Spp1 in atherosclerotic
plaques. PHD2cKO leads to apoptosis in TREM2hi macrophages.

## SessionInfo
```{r}
sessionInfo()

{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```
