---
title: "Role of PHD2 atherogenesis in in-vivo plaque Neutrophils"
author: "Javier Perales-PatÃ³n - javier.perales@bioquant.uni-heidelberg.de - ORCID: 0000-0003-0780-6683" 
output: github_document
---

> We have extensive data that supports the impact of PHD2 inhibition on promoting an aggravated 
atherogenesis phenotype via Macrophage apoptosis and its downstream signaling with Bnip3-Spp1 expression.
In-vivo model carries a PHD2cKO system dependent on Lyz2 expression. Macrophages but also neutrophil express
Lyz2 as main cell types in plaques. Thus it is unclear if the insights derived from macrophages could be
replicated by Neutrophils as both are effective of the PHD2cKO, potentially contributing both to the 
observed phenotype. Here we explore this contribution to discard a potential role of neutrophils in the
phenotype.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup
We define a random seed number for reproducibility, file structure for the output, 
and load essential libraries

### Environment
```{r env}
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./05_neutrophil_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png','tiff'))
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);
```

### Load libraries
```{r}
suppressPackageStartupMessages(require(Seurat))
suppressPackageStartupMessages(require(GSEABase))
suppressPackageStartupMessages(require(cowplot))
suppressPackageStartupMessages(require(genesorteR))
suppressPackageStartupMessages(require(fgsea))
suppressPackageStartupMessages(require(ggplot2))
suppressPackageStartupMessages(require(ggrepel))
suppressPackageStartupMessages(require(AUCell))
suppressPackageStartupMessages(require(viper))
suppressPackageStartupMessages(require(purrr))
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(require(openxlsx))
suppressPackageStartupMessages(require(ComplexHeatmap))
source("../src/graphics.R")
source("../src/seurat_fx.R")
source("../src/wilcox_fx.R")
```

## Load data
Read the Seurat Object from third step (Myeloid leukocyte map).
```{r load_seurat}
# Input data
sobj <- "./03_myeloidlineage_output/data/M.rds"
# Read 
if(file.exists(sobj)) {
	M <- readRDS(sobj)
} else {
	stop("ERROR: Seurat object does not exist. Run 03.rmd to generate it.")
}
```

We focus on Neutrophils in this analysis. So we get the cellIds for this population.
We show the sample size of neutrophils. It seems our data presents low representation of 
neutrophil population though.

```{r neutrophil_cellIds}
NeutroClasses <- grep("Neutro", levels(M), value=TRUE)
print(NeutroClasses)

NeutroIDs <- WhichCells(M, ident = NeutroClasses)

# Total cells
length(NeutroIDs)
# Sample size for each class
table("Class"=Idents(M[, NeutroIDs]))
# We save the sample size
write.table(as.data.frame(table("Class"=Idents(M[, NeutroIDs]))),
	    file=paste0(DATADIR,"/nCell_per_class.csv"),
	    sep=",", row.names=FALSE, col.names=TRUE, quote=FALSE)
```

## Comparison between PHD2 functional perturbation

>The in-vivo model carries a PHD2 knock-out in the myeloid lineage with an effectivity of ~30% (estimated)
of the cells. We have shown that PHD2cKO condition presents higher penetrance of BMDM-derived PHD2cKO signature.
The section will investigate if the penetrance of this gene perturbation is enough to be captured in the 
average population of the pooled PHD2cKO mice.

### Differential gene expression
We perform a standard differential expression with Bonferroni correction.

The scatter plot between the average expression of the two populations do NOT highlight Spp1
as one of the most differentially expressed genes in PHD2cKO condition in Neutrophils.

```{r scatterplot_neutrophils_PHD2cKO_vs_PHD2wt, fig.width=5, fig.height=5, dpi=300, warning=FALSE}
S1 <- M[, NeutroIDs] # Select neutrophils to pool for a test diff expr

Idents(S1) <- "stim"
avg <- log1p(AverageExpression(S1, verbose = FALSE)$RNA)
  
rnk <- setNames(avg[,"PHD2cKO"] - avg[,"WT"], rownames(avg))
top <- names(sort(abs(rnk),decreasing = TRUE)[1:15])
  
ggplot(avg, aes(WT, PHD2cKO)) + geom_point(alpha=0.6) + ggtitle(label = "Neutrophils") + 
	xlab("WT") + ylab("PHD2cKO")+
	geom_label_repel(data=avg[top,], aes(label=top), size=5) + 
	theme_cowplot() +
      theme(axis.title = element_text(size=22),
            title = element_text(size=22, hjust = 0.5),
            axis.text = element_text(size=18),
            axis.title.x = element_text(size=18),
            axis.title.y = element_text(size=18))
    
  
dge <- FindMarkers(object = S1, ident.1 = "PHD2cKO", ident.2 = "WT", 
		  logfc.threshold=0, min.pct=0)
cols_names <- colnames(dge)
# Add two extra cols
dge$cluster <- "Neutrophil PHD2cKO vs PHD2wt"
dge$gene <- rownames(dge)

dge <- dge[,c("cluster", "gene", cols_names)]
   
write.table(dge,
            file = paste0(DATADIR,"/Neutrophils_avg_PHD2cKO_vs_PHD2wt_DEGs.tsv"),
            sep="\t",col.names = TRUE, row.names = FALSE, quote=FALSE
)
```

We show the number of differentially expressed genes (Bonferroni-adjusted pvalue < 0.05):
```{r}
table(sign(dge$avg_logFC) * as.integer(dge$p_val_adj < 0.05))
```

> We could conclude that there are not large differences in the average gene expression of 
neutrophils based on condition (PHD2 KO). 

Next efforts will be made to dissect the main
transcriptional programme of this perturbation (Hypoxia) and explore the Bnip3-Spp1 axis
in neutrophils.

## Dissection of hypoxia response and PHD2 functional impairment

> We know that PHD2 is an important regulator in hypoxia response. We perform a
transcriptional dissection of hypoxia driven by the hypothesis that 
PHD2-KO perturbation triggers a strong hypoxia response. Thus it is expected that the 
PHD2-cKO condition presents an over-representation of cells responding to hypoxia.

### Dorothea focused on Hif1a
We calculate Hif1a transcription factor activities using dorothea. We know that Hif1a is 
the canonical transcription factor for hypoxia response, and PHD2 interacts with it to trigger
such a response.

```{r vln_neutrophils_dorothea_hif1a, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
# load regulons
df2regulon <- function(df, regulator_name="tf") {
  regulon = df %>% split(.[regulator_name]) %>% map(function(dat) {
    targets = setNames(dat$mor, dat$target)
    likelihood = dat$likelihood
    list(tfmode = targets, likelihood = likelihood)
  })
  return(regulon)
}

regulon.df <- read.table("../data/Prior/dorothea_regulon_mouse_v1.csv", sep=",", 
			 header=TRUE, stringsAsFactors = FALSE)
regul <- df2regulon(df=regulon.df)

# Calculate TF activities
TF <- viper(eset = as.matrix(M@assays$RNA@data), regulon = regul,
              nes = T, minsize = 4,
              eset.filter = F, adaptive.size = F,
	      verbose=FALSE)
  
# Add them as metadata
stopifnot(colnames(M) == colnames(TF))
M$Hif1a_activity <- TF["Hif1a",]
rm(TF)

# Visualization focused on Neutrophils population
VlnPlot.stim(M[, NeutroIDs], # Select Neutrophils
     meta.feature = "Hif1a_activity", ylab="Hif1a activity", 
     fontTXT=fontTXT)

```

### PROGENy focused on hypoxia response
We calculate a score of hypoxia response using progeny as well.

```{r vln_progeny_hypoxia_neutrophils, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
### Progeny ####
progeny.mat <- read.table("../data/Prior/progeny_matrix_mouse_v1.txt",sep=",",header=TRUE)
rownames(progeny.mat) <- progeny.mat$X
progeny.mat <- progeny.mat[which(colnames(progeny.mat)!="X")]
progeny.mat <- as.matrix(progeny.mat)

common <- intersect(rownames(M), rownames(progeny.mat))
  
prog <- t(as.matrix(M@assays$RNA@data[common,])) %*% progeny.mat[common,]
rn <- rownames(prog)
prog <- apply(prog,2,scale)
rownames(prog) <- rn
prog <- t(prog)
  
stopifnot(colnames(M) == colnames(prog))
M$Hypoxia_response <- prog["Hypoxia",]
rm(common,prog)

# Visualization focused on Neutrophils population
VlnPlot.stim(M[, NeutroIDs], # Neutrophils
     meta.feature = "Hypoxia_response", ylab="Hypoxia response (Score)", 
     fontTXT=fontTXT)
```

### Enrichment of in-vitro PHD2-KO signature
In the previous step, [03_myeloidlineage.md](03_myeloidlineage.md), we have already
calculated a single-cell score of the enrichment of in-vitro PHD2-KO signature among
myeloid leukocytes, including neutrophils.

> NOTE: Since we have already statistically tested this hypothesis among myeloid leukocytes,
we use the statistics from previous analysis so we do not have to correct for multiple
testing twice.

Here we show it again, but restricted to Neutrophils.

```{r vlnsplit_neutrophil_PHD2cKO, fig.width=3, fig.height=6, warning=FALSE}
wPHD2cKO_stats <- read.table(file=paste0("./03_myeloidlineage_output/data",
					 "/PHD2cKO_wilcox_stats.csv"),
			     sep=",", header=TRUE, stringsAsFactors=FALSE)
rownames(wPHD2cKO_stats) <- wPHD2cKO_stats$X

RES <- wPHD2cKO_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

stopifnot(all(levels(M)==rownames(RES)))
plotVln(SeuratObject = M[, grep("^Neutro", Idents(M))], 
	gene=NULL,meta="PHD2cKO",
	stats=RES[grep("^Neutro", rownames(RES)), ],
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.005, pt.alpha=0.4) +
ggtitle("PHD2cKO\nsignature")
```


### Pair-wise comparison of these markers of hypoxia response

```{r pairs_hypoxia_neutrophils, fig.width=9, fig.height=9, dpi=300, warning=FALSE}
# Define colors for ggplot
cond.cols <- c("WT"="grey", "PHD2cKO"="red")

panel.cor <- function(x, y, digits=2, prefix="", cex.cor, family=fontTXT, ...)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y, method="pearson"))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  txt <- paste(prefix, txt, sep="")
  if(missing(cex.cor)) cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r, family=family)
}

# Visualization
pairs(M@meta.data[NeutroIDs,
      		c("Hif1a_activity","Hypoxia_response","PHD2cKO")],
      lower.panel=panel.smooth, upper.panel=panel.cor, cex.labels = 2.6, cex.axis=2.3, las=1,
      family=fontTXT,
      pch=21,
      bg=cond.cols[M@meta.data[NeutroIDs, "stim"]])
```

> Conclusion: We could conclude that PHD2-cKO condition presents a skewed distribution of cells towards 
hypoxia response as compared to PHD2-wt condition. The PHD2cKO signature expression is highly correlated
to Hif1a transcription factor and hypoxia response activities. Their activity and expression is higher in
neutrophils from PHD2cKO condition. These observations are similar to the findings on Macrophages.

## PHD2KO-Bnip3-Spp1 axis

We already have shown that PHD2cKO signature is recovered in PHD2cKO condition. 
What about the Bnip3 and Spp1 expression? These has been already explored among
other myeloid leukocytes, but it is shown here focused on neutrophils. 

### Bnip3 expression across neutrophils

> NOTE: Since we have already statistically tested this hypothesis among myeloid leukocytes,
we use the statistics from previous analysis so we do not have to correct for multiple
testing twice.

```{r vlnsplit_neutrophil_Bnip3, fig.width=3, fig.height=6, warning=FALSE}
wBnip3_stats <- read.table(file=paste0("./03_myeloidlineage_output/data",
					 "/Bnip3_wilcox_stats.csv"),
			     sep=",", header=TRUE, stringsAsFactors=FALSE)
rownames(wBnip3_stats) <- wBnip3_stats$X

RES <- wBnip3_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

stopifnot(all(levels(M)==rownames(RES)))
vln_Bnip3 <- plotVln(SeuratObject = M[, grep("^Neutro", Idents(M))], 
		     gene="Bnip3",meta=NULL,
		     stats=RES[grep("^Neutro", rownames(RES)), ],
		     vlnsplit = TRUE, fontTXT,  nCell.y=-0.3, pt.alpha=0.4) 
vln_Bnip3
```


### Spp1 expression across neutrophils

```{r vlnsplit_neutrophil_Spp1, fig.width=3, fig.height=6, warning=FALSE}
wSpp1_stats <- read.table(file=paste0("./03_myeloidlineage_output/data",
					 "/Spp1_wilcox_stats.csv"),
			     sep=",", header=TRUE, stringsAsFactors=FALSE)
rownames(wSpp1_stats) <- wSpp1_stats$X

RES <- wSpp1_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

stopifnot(all(levels(M)==rownames(RES)))
vln_Spp1 <- plotVln(SeuratObject = M[, grep("^Neutro", Idents(M))], 
		    gene="Spp1",meta=NULL,
		    stats=RES[grep("^Neutro", rownames(RES)), ],
		    vlnsplit = TRUE, fontTXT,  nCell.y=-0.3, pt.alpha=0.4) 
vln_Spp1
```

> Conclusion: Bnip3 is hampered by drop-outs in single-cell profiling (zero-inflation), 
thus it makes difficult to compare groups. However, it seems that Bnip3 is up-regulated 
upon PHD2cKO. In contrast, PHD2cKO-induced up-regulation of Spp1 expression seems to be 
specific of Trem2-foamy neutrophils.

Next we stratify neutrophils to compare those undergoing complete PHD2-KO.

### Stratification of cells by BMDM-derived PHD2cKO signature
We just observed that the penetrance of PHD2-KO is not complete in the PHD2cKO condition.
Thus we use the expression of the in-vitro signature to compare those cells undergoing
PHD2 versus WT. For this, we choose an arbitrary threshold, the 3rd quartile of the signature
expression in PHD2cKO condition. We select these cells to be compared to WT neutrophils that are 
not responding to Hypoxia/PHD2cKO, by using the threshold of being below the Q3 cutoff. 

```{r}
cutoff <- quantile(M@meta.data[colnames(M) %in% NeutroIDs & M$stim=="PHD2cKO", "PHD2cKO"], 
		   probs=0.75) 

M$PHD2cKO_class <- NA

M$PHD2cKO_class[NeutroIDs] <- ifelse(M$PHD2cKO[NeutroIDs] > cutoff,
						   "High", "Low")
 
#NOTE: here it is where we concatenate condition + stratified class by PHD2cKO signature
# so later we could just compare the two groups of interest:
# - PHD2cKO neutrophils above Q3 signature expression (high)
# - WT neutrophils below Q3 signature expression (Low)
M$PHD2cKO_class2 <- paste0(M$stim,"_",M$PHD2cKO_class)

# Show cutoff used
print(cutoff)
# Save sample sizes
table(Strata=M$PHD2cKO_class2[NeutroIDs])
# We save the sample size
write.table(as.data.frame(table("Strata"=Idents(M[, NeutroIDs]))),
	    file=paste0(DATADIR,"/nCell_per_class.csv"),
	    sep=",", row.names=FALSE, col.names=TRUE, quote=FALSE)
```

The cutoff of Q3 in PHD2cKO is `r cutoff`.

For the records, we just show which cells has been selected below/above that line.

```{r vln_aucell_phd2ko_neutrophils_wCutoff, fig.width=6, fig.height=7, dpi=300, warning=FALSE}
# Visualization focused on Neutrophils population
VlnPlot.stim(M[,NeutroIDs], 
     meta.feature = "PHD2cKO", ylab="Expression of PHD2cKO signature (AUC)", 
     fontTXT=fontTXT) + geom_hline(yintercept=cutoff)
```

We retrieve only the neutrophils involved in the group of interest from now on. So the rest
of the analysis is focused on this stratification of cells

```{r sel_neutro}
NeutroIDs <- NeutroIDs[which(M$PHD2cKO_class2[NeutroIDs] %in% c("PHD2cKO_High", "WT_Low"))]
```

Perform a genome-wide differential gene expression between the two stratified groups.

```{r dge_PHD2cKO_High, warning=FALSE}
S1 <- M[, NeutroIDs] 
Idents(S1) <- "PHD2cKO_class2"
dge <- FindMarkers(S1, ident.1 = "PHD2cKO_High", ident.2 = "WT_Low",
		   min.pct = 0, logfc.threshold = 0)

saveRDS(dge, file="Neutro_PHD2cKO_High_DGE.Rds")
```

We generate a volcano plot to visually inspect the transcriptome-wide differences.
```{r volcano_PHD2cKOsign_high_vs_low, dpi=300, fig.width=8, fig.height=6, warning=FALSE}
Neutro_interesting_genes <- getGmt("../data/Prior/MC_PHD2_diff.gmt")
Neutro_interesting_genes <- unlist(geneIds(Neutro_interesting_genes))
if(any(!Neutro_interesting_genes %in% rownames(S1))) {
	cat(paste("The following genes are not present in the gene expr matrix:","\n",
		  paste(setdiff(Neutro_interesting_genes, rownames(S1)), collapse=","),
		  "\n"),
		  file=stdout())
	Neutro_interesting_genes <- intersect(Neutro_interesting_genes, rownames(S1))
}

dge$significant <- ifelse(dge$p_val_adj < 0.05,"Adj.P-value<0.05","n.s.")
dge$genes <- rownames(dge)
dge$show <- dge$genes %in% Neutro_interesting_genes
 
p = ggplot(dge, aes(avg_logFC, -log10(p_val+0.001))) +
    geom_point(aes(col=significant), alpha=0.6) +
    scale_color_manual(values=c("red", "black"))
  
  
p+geom_label_repel(data=dge[dge$show,],
                          aes(label=genes), family=fontTXT, size=8,
                          force=2,
			  xlim = c(2.5,3.0),
                          nudge_y=0.05, direction = "y", segment.size = 0.5) + 
    coord_cartesian(xlim = c(-1.3, 2.3), clip = "off") +
    theme_cowplot() +
    xlab("avg log2FC") + ylab("-log10(pvalue + 0.001)") +
    theme(text = element_text(family=fontTXT, size=20),
	  legend.text = element_text(family=fontTXT, size=20),
	  axis.text = element_text(family=fontTXT, size=20),
    	  plot.margin = unit(c(0.3, 2.6, 0.1, 0.1), "cm"),
	  legend.position = "bottom",
	  legend.direction = "horizontal"
	  )
```

We observe two points here:

- Bnip3 is slightly over-expressed in High PHD2cKO signature neutrophils. But not significant.
- Spp1 is not over-expressed in High PHD2cKO neutrophils. In opposite to Trem2-foamy Macrophages 
  from previous analysis, it is actually down-regulated upon PHD2cKO perturbation.


## Re-visit Bnip3 and Spp1 expression in stratified neutrophils
> We just distilled which neutrophils are undergoing PHD2-KO in the PHD2cKO condition
and which ones are WT sampled from WT condition. We compared their differential expression.
Now we re-visit a potential differential expression of Bnip3 and Spp1 following the evidences
from in-vitro experiments and data, but following a refined analysis with only these groups of 
neutrophils (high and low PHD2cKO signature).

```{r dotplot_Bnip3Spp1, fig.width=7, fig.height=4}
S1 <- M[, NeutroIDs]
S1$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S1$PHD2cKO_class2])
S1$stim <- relevel(S1$stim, ref=grep("WT", levels(S1$stim), value=TRUE))
cnt <- table(S1$stim, Idents(S1))>10
sel2 <- colnames(cnt)[colSums(cnt)==2]

DotPlot(S1[, WhichCells(S1, ident=sel2)], 
	features=c("Bnip3", "Spp1"), 
	split.by="stim", cols=c("blue","red"))
```

### Bnip3 in high PHD2cKO Neutrophils

```{r wilcox_bnip3_NeutroPHD2sign}
S1 <- M[, NeutroIDs]
S1$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S1$PHD2cKO_class2])
S1$stim <- relevel(S1$stim, ref=grep("WT", levels(S1$stim), value=TRUE))

wBnip3_stats <- t(sapply(levels(S1), function(cell) {
				   cellIds <- WhichCells(S1, idents=cell)
				   # The test
				   wilcox.test_stats(xy=S1@assays$RNA@data["Bnip3", cellIds],
						     gr=S1$stim[cellIds])
			   }))
wBnip3_stats <- as.data.frame(wBnip3_stats)
wBnip3_stats$adjpval <- p.adjust(wBnip3_stats$pvalue, method="fdr")
wBnip3_stats$significance <- tagSignif(wBnip3_stats$adjpval)
print(wBnip3_stats)
# Save it
write.table(wBnip3_stats, sep=",",
	    file=paste0(DATADIR,"/PHD2sign_Bnip3_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```

```{r vlnsplit_Bnip3_NeutroPHD2sign, fig.width=4, fig.height=6, warning=FALSE}
RES <- wBnip3_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = S1, gene="Bnip3", meta=NULL,
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.1, pt.alpha=0.4) +
      guides(fill= guide_legend(nrow=2, byrow = TRUE)) 
```


### Spp1 in high PHD2cKO Neutrophils

```{r wilcox_spp1_NeutroPHD2sign}
S1 <- M[, NeutroIDs]
S1$stim <- factor(c("WT_Low"="WT (Low signature)",
		    "PHD2cKO_High"="PHD2cKO (High signature)")[S1$PHD2cKO_class2])
S1$stim <- relevel(S1$stim, ref=grep("WT", levels(S1$stim), value=TRUE))

wSpp1_stats <- t(sapply(levels(S1), function(cell) {
				   cellIds <- WhichCells(S1, idents=cell)
				   # The test
				   wilcox.test_stats(xy=S1@assays$RNA@data["Spp1", cellIds],
						     gr=S1$stim[cellIds])
			   }))
wSpp1_stats <- as.data.frame(wSpp1_stats)
wSpp1_stats$adjpval <- p.adjust(wSpp1_stats$pvalue, method="fdr")
wSpp1_stats$significance <- tagSignif(wSpp1_stats$adjpval)
print(wSpp1_stats)
# Save it
write.table(wSpp1_stats, sep=",",
	    file=paste0(DATADIR,"/PHD2sign_Spp1_wilcox_stats.csv"),
	    row.names=TRUE, col.names=NA)
```

```{r vlnsplit_Spp1_NeutroPHD2sign, fig.width= 4, fig.height=6, warning=FALSE}
RES <- wSpp1_stats[, c("r", "significance")]
RES$r <- format(round(RES$r, 2), nsmall=2)
colnames(RES)[ncol(RES)] <- ""

plotVln(SeuratObject = S1, gene="Spp1", meta=NULL,
	stats=RES,
	vlnsplit = TRUE, fontTXT,  nCell.y=-0.3, pt.alpha=0.4) +
      guides(fill= guide_legend(nrow=2, byrow=TRUE)) 
```

## Conclusion

> We have dissected PHD2cKO perturbation in Neutrophils as we did with Macrophages. 
The PHD2cKO condition presents uncomplete but enriched penetrance of the perturbation, 
and we were able to recover which cells are undergoing the perturbation by using a BMDM-derived 
PHD2cKO signature from in-vitro experiments. Doing so, we were able to compare PHD2cKO and WT neutrophils 
in the two conditions, and stratify them by those most responding to the perturbation. 
We confirmed that PHD2cKO leads to hypoxia response.
We also found that Bnip3 is slightly up-regulated in PHD2cKO neutrophils (nonsignificant, p-value > 0.05), 
but NOT that strong and statistically significant as in Macrophages.
In opposite to Spp1 over-expression by Trem2-foamy macrophage upon PHD2cKO, Neutrophils do not over-express Spp1,
but even down-expressed it upon PHD2cKO (nonsignificant, p-value > 0.05).

## SessionInfo
```{r}
sessionInfo()

{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```
