---
title: "Exploratory Data Analysis (EDA) on conventional bulk RNA-seq from in-vitro Macrophage PHD2cKO and medium conditioned Fibroblasts"
author: "Javier Perales-Patón - javier.perales@bioquant.uni-heidelberg.de - ORCID: 0000-0003-0780-6683" 
output: github_document
---

Transcriptome profiling with conventional RNAseq was performed in three replicates of macrophages 
carrying PHD2 knock-out (with wildtype) and fibroblasts co-cultured with supernadant of these 
macrophages (as compared to supernadant of the wiltype genotype). Herein we perform an exploratory
analysis on these settings.

## Setting-up environment
The environment will be set with a random seed number for reproducibility and an output folder
for processed data and figures.

### set env
```{r env}
options(stringsAsFactors = FALSE)
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./00_EDA_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png','tiff'))
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);
```

### Load libraries
Essential libraries for R analysis.
```{r}
library(edgeR)
library(limma)
library(fgsea)
library(GSEABase)
library(dendextend)
source("../src/graphics.R")
```

## Load data and normalize
The data must be stored locally in the `./data/bulk` folder. 
This processed data has been deposited under the accession number 
described in the main [README file](../README.md) of the project.
```{r}
### 1 Load data
cnt <- read.table("../data/bulk/MST_MC_N_fibroblasts_PHD2-rawdata.txt", 
		  sep="\t", header = TRUE, check.names = FALSE)
rownames(cnt) <- cnt[,1]
cnt <- cnt[,-1]

targets <- read.table("../data/bulk/MST_MC_N_fibroblasts_PHD2-metadata.txt", 
		      sep="\t",header=TRUE, colClasses = "character")
targets$Sample <- gsub(" ","_",targets$Sample)

# Rename samples
stopifnot(all(colnames(cnt)==targets$FastQ))
colnames(cnt) <- targets$Sample

# Create metadata
stopifnot(all(colnames(cnt)==targets$Sample))
gr <- targets$Group
gt <- factor(gsub("^(3T3|MΦ)_", "", targets$Group), levels=c(c("PHD2_WT", "PHD2_KO")))
cell <- factor(unlist(sapply(targets$Group, function(z) strsplit(z, split="_")[[1]][1])))
repl <- factor(targets$Replicate)

## 2 Create DGElist (edgeR package, followed by TMM default normalization method)
y <- DGEList(counts = cnt, group = gr, genes = rownames(cnt))
y <- calcNormFactors(y, method = "TMM")
```

## Diagnostics
### The library size
```{r}
par(mar=c(10,8,4,4)); 
barplot(y$samples$lib.size,names.arg = rownames(y$samples), las=2)
```

### Principal Component analysis
First we start with the definition of ad-hoc handle functions to quick generate PCA plots 
(i.e. two-first principal components showing their variance explained).
```{r}
# Define some functions
runPCA <- function(mat) {
  pca <- prcomp(t(mat))
  
  return(pca)
}

plotPCA <- function(pca, dims, pchs, cols, labels=NULL, family="Arial") {
  importance <- summary(pca)$importance[,dims]
  PCscores <- data.frame(pca$x)[,dims]
  
  plot(PCscores,las=1,
       pch=pchs,bg=cols,
       family=family,
       cex.axis=1.3, cex.lab=1.3,
       xlab=paste0(dims[1]," : ",format(round(importance[2,dims[1]]*100,digits = 2),nsmall = 2),"% var expl."),
       ylab=paste0(dims[2]," : ",format(round(importance[2,dims[2]]*100,digits = 2),nsmall = 2),"% var expl."))
  if(!is.null(labels)) {
    par(xpd=TRUE)
    text(x=PCscores[,dims[1]], y=PCscores[,dims[2]], labels=labels, pos=1, family=family)
  }
}

```

Then we run PCA and visualize the PCA plots.
```{r pca, dpi=300, fig.width=7, fig.height=5}
# Run PCA
pca <- runPCA(mat = cpm(y, log = TRUE))

par(mar=c(4,4,4,12), xpd=TRUE)
plotPCA(pca=runPCA(mat = cpm(y, log = TRUE)),
        dims=c("PC1","PC2"),
        pchs=c(21,23)[as.integer(cell)], cols=as.integer(gt))
legend("right",legend = levels(cell), pch=c(21,23), 
       inset = c(-0.3,0), title = "Cell type")
legend("topright",legend = levels(gt), pch=c(22),pt.bg = c("black","red"), 
       inset = c(-0.4,0), title = "Perturbation")

# Which hallmarks of cell state are explained by those principal components
PC1_loadings <- sort(abs(pca$rotation[,"PC1"]),decreasing = TRUE)
GSC <- getGmt("../data/MSigDB/h.all.v6.2.symbols.gmt")
res <- fgsea(pathways = geneIds(GSC), stats = setNames(PC1_loadings, toupper(names(PC1_loadings))), nperm = 1000)
res <- res[order(res$padj,decreasing = FALSE),]

options(digits=2)
print(res[, 1:5])
```

We could conclude that PHD2 perturbation is the main factor of variability in the experiment.
Moreover, It seems Hypoxia response is involved in this separation. 
This makes sense given the knowledge on the role of PHD2.

In addition, we also look into individual PCA for each experiment.

```{r pca_macrophages, fig.width=5, fig.height=5, dpi=300}
# Macrophages
plotPCA(pca=runPCA(mat = cpm(y[,cell=="MΦ"], log = TRUE)),
        dims=c("PC1","PC2"),
        pchs=c(21), cols=as.integer(gt[cell=="MΦ"]))
```

```{r pca_fibroblasts, fig.width=5, fig.height=5, dpi=300}
# Fibroblasts
plotPCA(pca=runPCA(mat = cpm(y[,cell=="3T3"], log = TRUE)),
        dims=c("PC1","PC2"), family=fontTXT,
        pchs=c(21), cols=as.integer(gt[cell=="3T3"]))
```

We conclude that first principal component separates conditions in each experiment.

### Hierchical clustering
Finally, we perform Hierarchical clustering with the top 10% most variable genes to explore 
pair-wise sample similarities. We would expect that biological replicates would cluster together.

```{r hc, dpi=300, fig.width=6, fig.height=7}
cpm_norm <- cpm(y,log=TRUE)
rowise_var <- apply(cpm_norm,1,var)
genes_qt75 <- names(rowise_var[rowise_var>= quantile(rowise_var, 0.90)])
cpm_transform <- sweep(cpm_norm, 1, apply(cpm_norm, 1, median, na.rm=T))

d <- dist(t(cpm_transform[genes_qt75, ]))
hc <- hclust(d, method = "complete")
dend <- as.dendrogram(hc)

# Color the branches based on the clusters:
dend <- color_branches(dend, k=2) 
dend <- assign_values_to_branches_edgePar(dend=dend, value=c(1.5,1.5), edgePar="lwd")

# plot(dend, las=1, ylab="height")

# Manually match the labels, as much as possible, to the real classification of the flowers:
labels_colors(dend) <- as.integer(factor(setNames(gt, 
						  targets$Sample)[labels(dend)]))

par(mar=c(8,5,4,4))
plot(dend, las=1, ylab="", cex.axis=1.5)
title(ylab="height", cex.lab=1.5, line=+3.7)
```

We conclude that biological replicates and conditions are cluster as it would be expected given 
the experimental settings.

## Session info

```{r}
sessionInfo()

{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```

