Inference of Transcription Factor activity in Mieloid PHD2 deficiency in
atherogenesis
================
Javier Perales-Pat√≥n - <javier.perales@bioquant.uni-heidelberg.de> -
ORCID: 0000-0003-0780-6683

Herein we perform a Transcription Factor (TF) analysis on the role of
Myeloid-derived PHD2 in atherogenesis. For that, we use normalized data
from the differential gene expression analysis and estimate TF
activities using DoRothEA mouse regulons. See main README file for
details.

## Setup environment

The environment will be set with a random seed number for
reproducibility and an output folder for processed data and figures.

### set env

``` r
options(stringsAsFactors = FALSE)
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./03_TF_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png','tiff'))
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);
```

### Load libraries

Essential libraries for R analysis.

``` r
library(limma)
library(viper)
```

    ## Loading required package: Biobase

    ## Loading required package: BiocGenerics

    ## Loading required package: parallel

    ## 
    ## Attaching package: 'BiocGenerics'

    ## The following objects are masked from 'package:parallel':
    ## 
    ##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    ##     clusterExport, clusterMap, parApply, parCapply, parLapply,
    ##     parLapplyLB, parRapply, parSapply, parSapplyLB

    ## The following object is masked from 'package:limma':
    ## 
    ##     plotMA

    ## The following objects are masked from 'package:stats':
    ## 
    ##     IQR, mad, sd, var, xtabs

    ## The following objects are masked from 'package:base':
    ## 
    ##     anyDuplicated, append, as.data.frame, basename, cbind,
    ##     colnames, dirname, do.call, duplicated, eval, evalq, Filter,
    ##     Find, get, grep, grepl, intersect, is.unsorted, lapply, Map,
    ##     mapply, match, mget, order, paste, pmax, pmax.int, pmin,
    ##     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,
    ##     setdiff, sort, table, tapply, union, unique, unsplit, which,
    ##     which.max, which.min

    ## Welcome to Bioconductor
    ## 
    ##     Vignettes contain introductory material; view with
    ##     'browseVignettes()'. To cite Bioconductor, see
    ##     'citation("Biobase")', and for packages 'citation("pkgname")'.

``` r
library(purrr)
```

## Load data

We use the output data from DGE step (01). In particular, the limma
objects which contains gene expression, statistics and design of the
contrasts.

``` r
v <- readRDS("./01_DGE_output/data/v.rds")
eBay <- readRDS("./01_DGE_output/data/eBay.rds")
```

## Transcription factor analysis using DoRothEA

We read the mouse regulons for TF analysis.

``` r
# Function to transform a data.frame to viper's regulon 
# format using purrr. Ack. to CH. Holland.
df2regulon <- function(df, regulator_name="tf") {
  regulon = df %>% split(.[regulator_name]) %>% map(function(dat) {
    targets = setNames(dat$mor, dat$target)
    likelihood = dat$likelihood
    list(tfmode = targets, likelihood = likelihood)
  })
  return(regulon)
}

# Read dorothea mouse TF regulons
regulon.df <- read.table("../data/Prior/dorothea_regulon_mouse_v1.csv", 
             sep=",", header=TRUE, stringsAsFactors = FALSE)
# We shrink the analysis to those TF regulons with high-to-intermediate confidence
regulon.df <- subset(regulon.df, confidence %in% c("A","B","C")) # E are discarded
# Transform data.frame to viper regulon format
regul <- df2regulon(df=regulon.df)
```

The TF activities are estimated using `viper`. Similarly to gene-level
differential expression, contrasts performed using Empirical Bayes
method implemented in `limma`

``` r
TF = viper(eset = v$E, 
       regulon = regul,
           nes = T, minsize = 4,
           eset.filter = F, 
       adaptive.size = F, 
       verbose=FALSE)

TF.fit <- lmFit(TF, v$design)
TF.cont.mat <- makeContrasts("Fib_PHD2"=Fib_PHD2_KO - Fib_PHD2_WT,
                          "MC_PHD2"=MC_PHD2_KO - MC_PHD2_WT,
                          levels=v$design)
TF.fit <- eBayes(TF.fit)
TF.fit2 <- contrasts.fit(TF.fit, TF.cont.mat)
TF.eBay <- eBayes(TF.fit2)
```

## Broad statistics on diff TFs

Just check how many TFs are differentially active

``` r
diffTF_cnt <- apply(decideTests(TF.eBay),2, table)
print(diffTF_cnt)
```

    ##     Contrasts
    ##      Fib_PHD2 MC_PHD2
    ##   -1       19       5
    ##   0       176     277
    ##   1        90       3

## Diagnostics and save results

Write tables as supplementary data for the records and further
interpretation.

``` r
for(cont in colnames(TF.cont.mat)) {
  cat(paste0("Registering differential expression for ",cont,"\n"),
    file=stdout())
  topTab <- topTable(TF.eBay, coef=cont, number = Inf)
  # DEGs table
  write.table(topTab, file=paste0(DATADIR,cont,"_diffTF.tsv"),sep="\t",
                                  row.names=TRUE, col.names=NA, quote=FALSE)
  # Volcano plot
  plot(TF.eBay$coefficients[,cont], -log10(TF.eBay$p.value[,cont]),
       xlab="log2-fold-change", ylab="-log10(pvalue)", cex=0.7)
  # Histogram of p-vals
  hist(TF.eBay$p.value[,cont])
}
```

    ## Registering differential expression for Fib_PHD2

![](./03_TF_output//figures/diagnostics_diffTF-1.png)<!-- -->![](./03_TF_output//figures/diagnostics_diffTF-2.png)<!-- -->

    ## Registering differential expression for MC_PHD2

![](./03_TF_output//figures/diagnostics_diffTF-3.png)<!-- -->![](./03_TF_output//figures/diagnostics_diffTF-4.png)<!-- -->
Histogram of p-values looks good, and volcano plots as well. Certainly
Fibroblasts presents higher proportion of TF dysregulated from the
medium conditioned with Macrophages PHD2cKO.

## Hypothesis-driven test of Hif1a up-regulation in Macrophages PHD2cKO

Based on previous pathway analysis and literature, it would be expected
that actually Hif1a transcription factors presents high activity upon
PHD2 dysfunction. As it is confirmed
below:

``` r
MC_PHD2cKO_Hif1a <- topTable(TF.eBay, coef="MC_PHD2", number=Inf)["Hif1a",]
```

We could conclude that Hif1a is 1.4215989-fold times more active in
macrophages upon PHD2-KO (FDR=0.0050943).

## Session info

``` r
sessionInfo()
```

    ## R version 3.6.1 (2019-07-05)
    ## Platform: x86_64-pc-linux-gnu (64-bit)
    ## Running under: Ubuntu 18.04.3 LTS
    ## 
    ## Matrix products: default
    ## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
    ## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
    ## 
    ## locale:
    ##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
    ##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_US.UTF-8    
    ##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_US.UTF-8   
    ##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
    ##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
    ## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
    ## 
    ## attached base packages:
    ## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
    ## [8] base     
    ## 
    ## other attached packages:
    ## [1] purrr_0.3.2         viper_1.18.1        Biobase_2.44.0     
    ## [4] BiocGenerics_0.30.0 limma_3.40.6        rmarkdown_1.15     
    ## [7] nvimcom_0.9-82     
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] Rcpp_1.0.2         knitr_1.24         magrittr_1.5      
    ##  [4] splines_3.6.1      MASS_7.3-51.4      lattice_0.20-38   
    ##  [7] rlang_0.4.0        stringr_1.4.0      tools_3.6.1       
    ## [10] grid_3.6.1         xfun_0.9           KernSmooth_2.23-16
    ## [13] e1071_1.7-2        htmltools_0.3.6    class_7.3-15      
    ## [16] yaml_2.2.0         survival_2.44-1.1  digest_0.6.21     
    ## [19] Matrix_1.2-17      segmented_1.0-0    evaluate_0.14     
    ## [22] stringi_1.4.3      compiler_3.6.1     mixtools_1.1.0

``` r
{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```
